<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>ageas.Launch() &mdash; AGEAS  documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="ageas.Plot_Regulon()" href="plot_regulon.html" />
    <link rel="prev" title="Examples" href="../examples.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> AGEAS
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../api.html">API reference</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../examples.html">Examples</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">ageas.Launch()</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Mouse-Embryonic-Fibroblast(MEF)-vs-Embryonic-Stem-Cells(ESC)-example:">Mouse Embryonic Fibroblast(MEF) vs Embryonic Stem Cells(ESC) example:</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#For-using-raw-data">For using raw data</a></li>
<li class="toctree-l4"><a class="reference internal" href="#For-using-normalized-gene-counts">For using normalized gene counts</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#Carbon-Tetrachloride(CCl4)-induced-liver-fibrosis-example:">Carbon Tetrachloride(CCl4) induced liver fibrosis example:</a></li>
<li class="toctree-l3"><a class="reference internal" href="#And-then…">And then…</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="plot_regulon.html">ageas.Plot_Regulon()</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">AGEAS</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="../examples.html">Examples</a> &raquo;</li>
      <li>ageas.Launch()</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/notebooks/launch.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
.jp-RenderedHTMLCommon table,
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
.jp-RenderedHTMLCommon thead,
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
.jp-RenderedHTMLCommon tr,
.jp-RenderedHTMLCommon th,
.jp-RenderedHTMLCommon td,
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
.jp-RenderedHTMLCommon th,
div.rendered_html th {
  font-weight: bold;
}
.jp-RenderedHTMLCommon tbody tr:nth-child(odd),
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
.jp-RenderedHTMLCommon tbody tr:hover,
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="ageas.Launch()">
<h1>ageas.Launch()<a class="headerlink" href="#ageas.Launch()" title="Permalink to this heading"></a></h1>
<p>This notebook demonstrate how to use ageas.Launch() to launch AGEAS in extracting key genomics features from Gene Expression Matrices(GEMs) containing RNA-seq based gene counts data of different sample groups.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">ageas</span>
</pre></div>
</div>
</div>
<p>Currently, AGEAS support input data under two different formats:</p>
<ol class="arabic">
<li><p><strong>‘gem_files’</strong>: Gene expression of all groups are presented as dataframes under CSV or TXT format with rows representing genes and columns representing samples.</p>
<p>Example:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 23%" />
<col style="width: 15%" />
<col style="width: 15%" />
<col style="width: 15%" />
<col style="width: 15%" />
<col style="width: 15%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p></p></th>
<th class="head"><p>SRR1039509</p></th>
<th class="head"><p>SRR1039512</p></th>
<th class="head"><p>SRR1039513</p></th>
<th class="head"><p>SRR1039516</p></th>
<th class="head"><p>SRR1039508</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>ENSG00000000003</p></td>
<td><p>679</p></td>
<td><p>448</p></td>
<td><p>873</p></td>
<td><p>408</p></td>
<td><p>1138</p></td>
</tr>
<tr class="row-odd"><td><p>ENSG00000000005</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
</tr>
<tr class="row-even"><td><p>ENSG00000000419</p></td>
<td><p>467</p></td>
<td><p>515</p></td>
<td><p>621</p></td>
<td><p>365</p></td>
<td><p>587</p></td>
</tr>
<tr class="row-odd"><td><p>ENSG00000000457</p></td>
<td><p>260</p></td>
<td><p>211</p></td>
<td><p>263</p></td>
<td><p>164</p></td>
<td><p>245</p></td>
</tr>
<tr class="row-even"><td><p>ENSG00000000460</p></td>
<td><p>60</p></td>
<td><p>55</p></td>
<td><p>40</p></td>
<td><p>35</p></td>
<td><p>78</p></td>
</tr>
<tr class="row-odd"><td><p>ENSG00000000938</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>2</p></td>
<td><p>0</p></td>
<td><p>1</p></td>
</tr>
</tbody>
</table>
<p>Genes must either be named with official gene symbols or Ensembl gene IDs(ENS***).</p>
<p>There is no requirement for sample name type. Barcodes, numbers, any artificial names can work.</p>
</li>
<li><p><strong>‘mex_folders’</strong>: Each folder containing Market Exchange Format (MEX) files output by <a class="reference external" href="https://github.com/10XGenomics/cellranger">cellranger</a> pipeline representing a group of samples. For more information: <a class="reference external" href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/output/matrices">https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/output/matrices</a></p></li>
</ol>
<div class="section" id="Mouse-Embryonic-Fibroblast(MEF)-vs-Embryonic-Stem-Cells(ESC)-example:">
<h2>Mouse Embryonic Fibroblast(MEF) vs Embryonic Stem Cells(ESC) example:<a class="headerlink" href="#Mouse-Embryonic-Fibroblast(MEF)-vs-Embryonic-Stem-Cells(ESC)-example:" title="Permalink to this heading"></a></h2>
<p>This example represent how to use ageas.Launch() with <strong>‘gem_files’</strong>.</p>
<p>Here, we attempt to extract key genomics factors to perform cell reprogramming from <strong>MEF</strong> into <strong>Induced Pluripotent Stem Cell(iPSC)</strong>, one of the most well known cell reprogramming case, with AGEAS.</p>
<p>The scRNA-seq based gene expression data for both <strong>MEF</strong> and <strong>ESC</strong> are retrieved from <a class="reference external" href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE103221">GSE103221</a>.</p>
<p>Either raw data in <strong>GSE103221_RAW.tar</strong> or normalized counts in <strong>GSE103221_normalized_counts.csv.gz</strong> can be processed with AGEAS.</p>
<div class="section" id="For-using-raw-data">
<h3>For using raw data<a class="headerlink" href="#For-using-raw-data" title="Permalink to this heading"></a></h3>
<p>With all setting remain default, we can launch AGEAS as:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ageas</span><span class="o">.</span><span class="n">Launch</span><span class="p">(</span>
    <span class="c1"># since &#39;gem_files&#39; is the default setting, this line can be deleted</span>
    <span class="n">database_type</span> <span class="o">=</span> <span class="s1">&#39;gem_files&#39;</span><span class="p">,</span>

    <span class="c1"># ageas.Data_Preprocess() args</span>
    <span class="n">group1_path</span> <span class="o">=</span> <span class="s1">&#39;GSE103221_RAW/GSM3629847_10x_osk_mef.csv.gz&#39;</span><span class="p">,</span>
    <span class="n">group2_path</span> <span class="o">=</span> <span class="s1">&#39;GSE103221_RAW/GSM3629848_10x_osk_esc.csv.gz&#39;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<p><strong>However, section above could be too computational expensive for PC!</strong></p>
<p>Few adjustments can be made:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">test_raw</span> <span class="o">=</span> <span class="n">ageas</span><span class="o">.</span><span class="n">Launch</span><span class="p">(</span>
    <span class="n">protocol</span> <span class="o">=</span> <span class="s1">&#39;multi&#39;</span><span class="p">,</span>
    <span class="n">unit_num</span> <span class="o">=</span> <span class="mi">4</span><span class="p">,</span>

    <span class="c1"># ageas.Data_Preprocess() args</span>
    <span class="n">group1_path</span> <span class="o">=</span> <span class="s1">&#39;GSE103221_RAW/GSM3629847_10x_osk_mef.csv.gz&#39;</span><span class="p">,</span>
    <span class="n">group2_path</span> <span class="o">=</span> <span class="s1">&#39;GSE103221_RAW/GSM3629848_10x_osk_esc.csv.gz&#39;</span><span class="p">,</span>
    <span class="n">std_value_thread</span> <span class="o">=</span> <span class="mf">3.0</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<p>With <strong>std_value_thread = 3.0</strong>, more genes with relatively low expression variability will be ruled out during meta-processing and, thus, limit amount of Gene Regulatory Pathways(GRPs) in meta level processed Gene Regulatory Network(GRN) and pseudo-sample GRNs.</p>
<p>Instead of using only one AGEAS extractor units by default, four units are used with <strong>unit_num = 4</strong>. Result now is generalized from GRPs extracted with every unit.</p>
<p>To reduce running time, <strong>protocol = ‘multi’</strong> set units to run parallelly with multithreading.</p>
<p>For more API information, please visit <a class="reference external" href="https://nkmtmsys.github.io/Ageas/html/generated/ageas.Launch.html#ageas.Launch">documentaion page</a>.</p>
<p>Extraction reports can be saved as files with:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">test_raw</span><span class="o">.</span><span class="n">save_reports</span><span class="p">(</span>
    <span class="n">folder_path</span> <span class="o">=</span> <span class="s1">&#39;report_files/&#39;</span><span class="p">,</span>
    <span class="n">save_unit_reports</span> <span class="o">=</span> <span class="kc">True</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<p>Within folder <em>report_files</em>, there should have following files:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>report_files/
    │
    ├─ no_1/
    │  ├─ grps_importances.txt
    │  ├─ outlier_grps.js
    │
    ├─ no_2/
    <span class="p">|</span>  ├─ ...
    │
    ├─ no_3/
    <span class="p">|</span>  ├─ ...
    │
    ├─ no_4/
    <span class="p">|</span>  ├─ ...
    │
    ├─ full_atlas.js
    ├─ grp_scores.csv
    ├─ key_atlas.js
    ├─ meta_GRN.js
    ├─ meta_report.csv
    ├─ pseudo_sample_GRNs.js
    ├─ report.csv
</pre></div>
</div>
<p>Folders <strong>no_1</strong>, <strong>no_2</strong>, <strong>no_3</strong>, <strong>no_4</strong> contain GRP importance scores returned by each extractor unit as <strong>grps_importances.csv</strong> which has GRPs ranked with importance scores and <strong>outlier_grps.js</strong> which has GRPs once removed during feature selection due to extremly high importance score. If these information not needed, keep <strong>save_unit_reports</strong> as <strong>False</strong> by default.</p>
<p><strong>full_atlas.js</strong>: regulons reconstructed with every important GRP extracted by every extractor unit.</p>
<p><strong>grp_scores.csv</strong>: GRPs with max importance score each can achieve after extracted by every extractor unit.</p>
<p><strong>key_atlas.js</strong>: pruned regulons reconstructed only with genes being capable to regulate other genes in full atlas.</p>
<p><strong>meta_GRN.js</strong>: meta-level processed GRN cast with all data in each sample group.</p>
<p><strong>meta_report.csv</strong>: summary of every gene in meta-processed GRN. By default, records are ranked by <strong>Degree</strong>. Top few rows should look like:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 14%" />
<col style="width: 26%" />
<col style="width: 9%" />
<col style="width: 14%" />
<col style="width: 37%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Gene Symbol</p></th>
<th class="head"><p>Type</p></th>
<th class="head"><p>Degree</p></th>
<th class="head"><p>Log2FC</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Pou5f1</p></td>
<td><p>Pou5f1</p></td>
<td><p>TF</p></td>
<td><p>786</p></td>
<td><p>18.0654266535883</p></td>
</tr>
<tr class="row-odd"><td><p>Trim28</p></td>
<td><p>Trim28</p></td>
<td><p>TF</p></td>
<td><p>727</p></td>
<td><p>16.7739633684336</p></td>
</tr>
<tr class="row-even"><td><p>Trp53</p></td>
<td><p>Trp53</p></td>
<td><p>TF</p></td>
<td><p>725</p></td>
<td><p>15.9708922902521</p></td>
</tr>
<tr class="row-odd"><td><p>Rest</p></td>
<td><p>Rest</p></td>
<td><p>TF</p></td>
<td><p>695</p></td>
<td><p>15.0240141813129</p></td>
</tr>
<tr class="row-even"><td><p>Sox2</p></td>
<td><p>Sox2</p></td>
<td><p>TF</p></td>
<td><p>687</p></td>
<td><p>15.4459524481466</p></td>
</tr>
<tr class="row-odd"><td><p>Junb</p></td>
<td><p>Junb</p></td>
<td><p>TF</p></td>
<td><p>683</p></td>
<td><p>14.1196706687477</p></td>
</tr>
<tr class="row-even"><td><p>Cebpb</p></td>
<td><p>Cebpb</p></td>
<td><p>TF</p></td>
<td><p>682</p></td>
<td><p>13.8599229728618</p></td>
</tr>
</tbody>
</table>
<ul class="simple">
<li><p>Type can either be Gene or TF(Transcription Factor).</p></li>
<li><p>Degree here stands for <a class="reference external" href="https://en.wikipedia.org/wiki/Degree_(graph_theory)">degree in graph theory</a>.</p></li>
<li><p>Log2FC is calcualted with all expression values in two given sample groups.</p></li>
</ul>
<p><strong>pseudo_sample_GRNs.js</strong>: GRNs cast for each pseudo-sample generated with <a class="reference external" href="https://stackoverflow.com/questions/8269916/what-is-sliding-window-algorithm-examples">Sliding Window Algorithm</a>.</p>
<p><strong>report.csv</strong>: information about key regulatory-source genes in <strong>full_atlas.js</strong>. By default, records are ranked by <strong>Log2FC</strong>. Top few rows should look like:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 8%" />
<col style="width: 14%" />
<col style="width: 11%" />
<col style="width: 5%" />
<col style="width: 13%" />
<col style="width: 13%" />
<col style="width: 14%" />
<col style="width: 23%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Gene Symbol</p></th>
<th class="head"><p>Regulon</p></th>
<th class="head"><p>Type</p></th>
<th class="head"><p>Source_Num</p></th>
<th class="head"><p>Target_Num</p></th>
<th class="head"><p>Meta_Degree</p></th>
<th class="head"><p>Log2FC</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Pou5f1</p></td>
<td><p>Pou5f1</p></td>
<td><p>regulon_0</p></td>
<td><p>TF</p></td>
<td><p>5</p></td>
<td><p>68</p></td>
<td><p>786</p></td>
<td><p>18.065426653588275</p></td>
</tr>
<tr class="row-odd"><td><p>Klf2</p></td>
<td><p>Klf2</p></td>
<td><p>regulon_0</p></td>
<td><p>TF</p></td>
<td><p>0</p></td>
<td><p>71</p></td>
<td><p>592</p></td>
<td><p>17.77230812962195</p></td>
</tr>
<tr class="row-even"><td><p>Trim28</p></td>
<td><p>Trim28</p></td>
<td><p>regulon_0</p></td>
<td><p>TF</p></td>
<td><p>8</p></td>
<td><p>136</p></td>
<td><p>727</p></td>
<td><p>16.77396336843356</p></td>
</tr>
<tr class="row-odd"><td><p>Trp53</p></td>
<td><p>Trp53</p></td>
<td><p>regulon_0</p></td>
<td><p>TF</p></td>
<td><p>7</p></td>
<td><p>133</p></td>
<td><p>724</p></td>
<td><p>15.97089229025213</p></td>
</tr>
<tr class="row-even"><td><p>Sox2</p></td>
<td><p>Sox2</p></td>
<td><p>regulon_0</p></td>
<td><p>TF</p></td>
<td><p>3</p></td>
<td><p>83</p></td>
<td><p>687</p></td>
<td><p>15.44595244814661</p></td>
</tr>
<tr class="row-odd"><td><p>Nanog</p></td>
<td><p>Nanog</p></td>
<td><p>regulon_0</p></td>
<td><p>TF</p></td>
<td><p>4</p></td>
<td><p>81</p></td>
<td><p>656</p></td>
<td><p>15.417885470810674</p></td>
</tr>
<tr class="row-even"><td><p>Klf4</p></td>
<td><p>Klf4</p></td>
<td><p>regulon_0</p></td>
<td><p>TF</p></td>
<td><p>3</p></td>
<td><p>62</p></td>
<td><p>428</p></td>
<td><p>15.348589852448784</p></td>
</tr>
</tbody>
</table>
<ul class="simple">
<li><p>Type can either be Gene or TF(Transcription Factor).</p></li>
<li><p>Source-Num and Target_Num indicate amount of directly related regulatory source and target in full atlas.</p></li>
<li><p>Meta_Degree and Log2FC here are same with Degree and Log2FC in <strong>meta_report.csv</strong>.</p></li>
</ul>
</div>
<div class="section" id="For-using-normalized-gene-counts">
<h3>For using normalized gene counts<a class="headerlink" href="#For-using-normalized-gene-counts" title="Permalink to this heading"></a></h3>
<p>We will need to make new gene expression matrices with all MEF samples and ESC samples respectively.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;GSE103221_normalized_counts.csv&#39;</span><span class="p">,</span> <span class="n">index_col</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>

<span class="n">mef_samples</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">data</span> <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;mef&#39;</span><span class="p">,</span> <span class="n">x</span><span class="p">)]</span>
<span class="n">esc_samples</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">data</span> <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;esc&#39;</span><span class="p">,</span> <span class="n">x</span><span class="p">)]</span>

<span class="n">data</span><span class="p">[</span><span class="n">mef_samples</span><span class="p">]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;mef.csv.gz&#39;</span><span class="p">)</span>
<span class="n">data</span><span class="p">[</span><span class="n">esc_samples</span><span class="p">]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;esc.csv.gz&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Considering the sample size of raw data for each group is few thousands, generating pseudo-samples, which abstracts gene expressions from several distinct samples as continuous expression data in order to calculate gene expression correlations, with every distinct 100 samples by default setting should be acceptable.</p>
<p>However, in this normalized data, only dozens of samples clearly labeled as MEF or ESC. To make at least dozens of pseudo-samples, we can adjust few arguments for Sliding Window Algorithm.</p>
<p>Furthermore, with normalized expression value, GRP filters shall also be adjusted to keep total amount of GRP after meta-processing reasonable. <strong>log2fc_thread</strong> can rule out genes and corresponding GRPs based on expression difference.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">test_normalized</span> <span class="o">=</span> <span class="n">ageas</span><span class="o">.</span><span class="n">Launch</span><span class="p">(</span>
    <span class="n">unit_num</span> <span class="o">=</span> <span class="mi">4</span><span class="p">,</span>

    <span class="c1"># ageas.Data_Preprocess() args</span>
    <span class="n">group1_path</span> <span class="o">=</span> <span class="s1">&#39;mef.csv.gz&#39;</span><span class="p">,</span>
    <span class="n">group2_path</span> <span class="o">=</span> <span class="s1">&#39;esc.csv.gz&#39;</span><span class="p">,</span>
    <span class="n">log2fc_thread</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span>
    <span class="n">std_value_thread</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span>
    <span class="n">sliding_window_size</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
    <span class="n">sliding_window_stride</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
<span class="p">)</span>

<span class="n">test_normalized</span><span class="o">.</span><span class="n">save_reports</span><span class="p">(</span>
    <span class="n">folder_path</span> <span class="o">=</span> <span class="s1">&#39;report_files/&#39;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<p>Files within folder <em>report_files</em> are under same structure and formats described above.</p>
</div>
</div>
<div class="section" id="Carbon-Tetrachloride(CCl4)-induced-liver-fibrosis-example:">
<h2>Carbon Tetrachloride(CCl4) induced liver fibrosis example:<a class="headerlink" href="#Carbon-Tetrachloride(CCl4)-induced-liver-fibrosis-example:" title="Permalink to this heading"></a></h2>
<p>This example represent how to use ageas.Launch() with <strong>‘mex_folders’</strong>.</p>
<p>Stimulating portal fibroblasts(PFs) as activated hepatic stellate cells(HSCs), we can try to find key genomics factors in liver fibrosis through extracting key genomic differences among HSCs and PFs.</p>
<p>The scRNA-seq based gene expression data for PF is retrieved from <a class="reference external" href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM4085627">GSM4085627</a> and data for HSC is retrieved from <a class="reference external" href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM4085625">GSM4085625</a>.</p>
<p>Files retrieved are managed as:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>a6w_pf/
    ├─ GSM4085627_10x_5_barcodes.tsv.gz
    ├─ GSM4085627_10x_5_genes.tsv.gz
    ├─ GSM4085627_10x_5_matrix.mtx.gz

a6w_hsc/
    ├─ GSM4085625_10x_3_barcodes.tsv.gz
    ├─ GSM4085625_10x_3_genes.tsv.gz
    ├─ GSM4085625_10x_3_matrix.mtx.gz
</pre></div>
</div>
<p>Then, we can launch AGEAS with:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">test_raw</span> <span class="o">=</span> <span class="n">ageas</span><span class="o">.</span><span class="n">Launch</span><span class="p">(</span>
    <span class="n">unit_num</span> <span class="o">=</span> <span class="mi">4</span><span class="p">,</span>

    <span class="c1"># ageas.Data_Preprocess() args</span>
    <span class="n">database_type</span> <span class="o">=</span> <span class="s1">&#39;mex_folders&#39;</span><span class="p">,</span>
    <span class="n">group1_path</span> <span class="o">=</span> <span class="s1">&#39;a6w_hsc/&#39;</span><span class="p">,</span>
    <span class="n">group2_path</span> <span class="o">=</span> <span class="s1">&#39;a6w_pf/&#39;</span><span class="p">,</span>
<span class="p">)</span>

<span class="n">test_normalized</span><span class="o">.</span><span class="n">save_reports</span><span class="p">(</span>
    <span class="n">folder_path</span> <span class="o">=</span> <span class="s1">&#39;report_files/&#39;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<p>Files within folder <em>report_files</em> are under same structure and formats described in example above.</p>
</div>
<div class="section" id="And-then…">
<h2>And then…<a class="headerlink" href="#And-then…" title="Permalink to this heading"></a></h2>
<p>We can visualize each regulon as a network graph with <a class="reference external" href="https://nkmtmsys.github.io/Ageas/html/generated/ageas.Plot_Regulon.html">ageas.Plot_Regulon()</a></p>
</div>
</div>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../examples.html" class="btn btn-neutral float-left" title="Examples" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="plot_regulon.html" class="btn btn-neutral float-right" title="ageas.Plot_Regulon()" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, JackSSK, nkmtmsys.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>