\documentclass[fleqn,10pt]{wlscirep}
\usepackage[utf8]{inputenc}
\usepackage[T1]{fontenc}
\usepackage{listings}
\usepackage{algorithm2e}
\usepackage{subcaption}
\usepackage[labelformat=parens,labelsep=quad,skip=3pt]{caption}
\usepackage{graphicx}

\title{AGEAS: Automated Machine Learning based Genetic Regulatory Element Extraction System}


% \author[1,2,*1]{Jack Yu}
\author[1,+]{Masayoshi Nakamoto}
\author[2,*1,+]{Jiawang Tao}
\author[1,2,*2,+]{Jack Yu}

% \author[2,+]{Derek Author}
\affil[1]{Shenzhen Mozhou Technology Co., Ltd, Shenzhen, China}
\affil[2]{Center for Health Research, Guangzhou Institutes of Biomedicine and Health, Chinese Academy of Sciences, Guangzhou 510530, China}
% \affil[2]{Affiliation, department, city, postcode, country}
\affil[*1]{Correspondence: tao\_jiawang@gibh.ac.cn}
\affil[*2]{Correspondence: gyu17@alumni.jh.edu}

\affil[+]{these authors contributed equally to this work}
\keywords{Machine Learning, Gene Regulatory Network, RNA-seq, Chip-seq}

\begin{abstract}
  As rapid progress in sequencing technology since last decade, numerous mechanisms underlying cell functions and developmental processes have been revealed as complex regulations of gene expressions.
  Since single-cell RNA sequencing (scRNA-seq) made high-resolution transcriptomic view increasingly accessible, precise identification of gene regulatory network (GRN) describing cell types and cell states became achievable.
  However, extracting key regulatory elements, including gene regulatory pathways (GRPs), transcription factors (TFs), and targetomes, that accurately and completely reflects functionality changes in biological phenomena remains challenging.
  Herein, we describe AGEAS, an semi-supervised automated machine learning (AutoML) based genetic regulatory element extraction system that assesses importances of GRPs in resulting biological phenomena, such as cell type differentiation, physiological and pathological development, and reconstructs GRNs with extracted important GRPs for comprehensive inference.
  With several case studies in divergent research areas, we show that AGEAS can indeed extract informative regulatory elements and reconstruct networks to indicate regulatory changes in biological phenomena of interest.
  % Furthermore, we validated applicability of AGEAS with empirical evidence that LIM homeobox 2 (Lhx2), key quiescent hepatic stellate cell (HSC) factor AGEAS extracted through analyzing chronic carbon tetrachloride ($CCl_4$) induced liver fibrosis data, can indeed halt HSC activation to prevent fibrogenic process.

  \subsection*{Availability and implementation}
  The AGEAS code is available at \url{https://github.com/JackSSK/Ageas}.
\end{abstract}

\begin{document}
\flushbottom
\maketitle

\thispagestyle{empty}
% \noindent Please note: Abbreviations should be introduced at the first mention in the main text â€“ no abbreviations lists. Suggested structure of main text (not enforced) is provided below.

\section*{Introduction}
  \label{introduction}
  As high-resolution sequencing technologies become increasingly applicable and accessible, an efficient and robost analytical system capable of extracting key genetic features responsible for cell types and cell states difference with limited prior biological knowledge also become highly demanded.
  Several methodologies like SCENIC \cite{Scenic} already demonstrated informativeness and robustness in studying cellular phenotypes with GRN analysis.
  Regulons, collections of a TF and corresponding targetomes, can successfully address cell identities with more comprehensive information compared with differential expressing genes (DEGs). \cite{Scenic, pySCENIC}
  Even though limited computational methods are capable of completely and precisly extracting key genetic regulatory elements in biological process of interest, Mogrify\cite{mogrify_2016} and CellNet\cite{cellnet_2014} have demonstrated GRN based methods can help to analyze cell type differentiation and to implement cell reprogramming.
  However, both methods may require large scale of additional background data to be applicable in other studies such as analyzing physiological or pathological development of selected cell type.
  The scREMOTE \cite{scREMOTE} published in 2022 is a potential method to extract generalized regulatory elements since limited background data is required with sequencing data representing sample classes.
  But prior knowledge on key TFs and marker genes is indispensable.
  A regulation trajectory of DEGs or regulons showing significant activity difference after GRN reconstruction would be able to reveal the developmental pathways as generalized approach.
  Due to general gene regulatory nature of TFs, numerous noisy signals are expected.
  Therefore, here we present AGEAS, an algorithm with feasibility in extracting key regulatory elements for biological phenomena of interest and can potentially promote further discoveries, even with scarce prior knowledge.

  Several reasons motivated us to develop semi-supervised AutoML based method.
  Firstly, accurate labeling of regulatory elements' relatedness in biological phenomena would require extensive prior knowledge may be unavailable.
  While unsupervised clustering can potentially extract factors playing important roles, this approach may focus on specific aspects with regulatory elements showing similar patterns and result in losing comprehensiveness.
  As a result, we chose to apply biological phenomena associating sample classes, such as cell types and cell states which should be easily retrievable, as labels for classification models to predict with masked GRN as input and then interpret success predictions to extract key regulatory elements.
  Secondly, it is also hard to guarantee comprehensiveness when developing a generalized classification model in differentiating GRNs, considering that capturing few significant differences would be sufficient for the model to reach outstanding accuracy.
  Hence, an extraction system with multiple independent classification models implemented with various algorithms should be helpful to recover from potential comprehensiveness loss.
  Thirdly, the process of interpreting all success prediction made by every classification model and integrating corresponding interpretation results may be computational expensive.
  Thus, applying an AutoML-based model and feature selection procedure that can effectively decrease interpretation cost would be necessary to increase feasibility of the method.

  To evaluate the performance of AGEAS, we primarily applied it on studying somatic cell reprogramming process and further addressed performance in analyzing 3 other biological processes in cell subtype differentiation, physiological and pathological development.
  % Later, we also addressed the applicability of AGEAS extraction result in guiding further studies about biological phenomena with $in$ $vitro$ experiments in determining Lhx2, key TF regulating liver fibrosis extracted with AGEAS, can indeed inhibit HSC activation.

\section*{Method}
  \label{method}
  The basic principle of AGEAS is to find key regulatory elements associated with biological phenomena of interest through analyzing how well-performing classification models distinguish GRNs of sample with the phenomena from those without.
  To reconstruct sufficient GRNs for each class, RNA-seq based expression data is segmented into subsets while each one is analogized as a pseudo-sample having discrete expression data.
  The pseudo-sample GRNs (psGRNs) are reconstructed accordingly; thus, classification models can be trained, evaluated, and interpreted with GRPs as input factors.
  With heavily weighted GRPs and corresponding genes repeatedly obtained from interpretations of successful sample class predictions, GRNs could be formed and potentially play an important role in differentiating the studying sample classes.
  The overall workflow can be summarized in Figure \ref{workflow}.
  By default, four separate extractor units in workflow run in parallel after data preprocessing part in order to increase output stability considering the stochastic nature of AGEAS, and all extracted regulatory elements are used to form GRNs, which will later be combined into one atlas.
  The following sections describe each step of AGEAS in more depth.
  % \begin{itemize}
  %   \item \hyperref[step1]{\textbf{\emph{Step 1}}}: Data preprocessing
  %   \item \hyperref[step2]{\textbf{\emph{Step 2}}}: Classification model selection.
  %   \item \hyperref[step3]{\textbf{\emph{Step 3}}}: Feature selection...
  %   \item \hyperref[step4]{\textbf{\emph{Step 4}}}: Top GRP extraction.
  %   \item \hyperref[step5]{\textbf{\emph{Step 5}}}: Key network reconstruction.
  % \end{itemize}


  \subsection*{Step 1:Data preprocessing}
    \label{step1}
    The main purpose of this step is to build pseudo-samples and to reconstruct corresponding pseudo-sample GRNs (psGRNs).
    For each sample class, gene expression matrices (GEMs) with the same class label are concatenated as one comprehensive expression matrix.
    With comprehensive GEMs, a meta-level GRN (meta-GRN) is reconstructed before psGRNs to provide generic guidance on reconstruction.
    In general, the workflow of this step can be summarized in Figure \ref{data_preprocess}.

    \subsubsection*{Reconstruct meta-GRN}
      Firstly, genes included in the comprehensive GEMs are assessed and determined whether having potential to form informative GRPs with other genes.
      Commonly, DEGs are considered as important factors of studying phenomena.
      Here we apply the Mann-Whitney U rank test (MWW) implemented by \emph{SciPy}\cite{2020SciPy-NMeth} to exclude genes with indistinguishable expression level distribution across GEMs of different classes.
      The p-value for rejecting null hypothesis, that expression profile underlying class 1 samples is the same as the expression profile underlying class 2 samples, is set to 0.05 by default.
      Furthermore, a $\log_2$ fold change (log2FC) filter is implemented in AGEAS.
      However, enabling the log2FC filter is not encouraged, considering that upstream TFs indirectly regulate key genes associated with phenomena of interest may not be significantly differential expressed.
      The log2FC filter shall mostly be used to decrease meta-GRN's total degree in a compromising position caused by limited computational resources.
      After differential expression based filters, a standard deviation ($\sigma$) filter is applied to exclude genes with low expression level or merely affected by dynamic expression status of other genes.
      By default, the $\sigma$ threshold is set to 1.0, the lowest expression level in raw gene count matrix gained from RNA-seq data.
      The threshold value should be adjusted based on prior knowledge of input GEMs, for example, the normalization method applied to the GEMs.

      With candidate genes passing filters above, some gene pairs are formed and evaluated by the potential of representing GRPs.
      To reduce overall computational complexity, a gene pair shall be formed with at least one TF, which could be the regulatory source of GRP.
      If not further specified, a TF list will be retrieved from integrated \emph{TRANSFAC}\cite{transfac} dataset according to the provided species information.
      Utilizing genetic interaction database like \emph{GTRD}\cite{gkaa1057} and \emph{BioGRID}\cite{biogrid}, AGEAS checks whether the binding ability of gene pair is confirmed or not.
      By default, if the TF recorded has at least one binding site within target gene's promoter range (-1000 to +100) by Chromatin Immunoprecipitation Sequencing (ChIP-seq) dataset retrieved from \emph{GTRD}\cite{gkaa1057}, the potential GRP gene pair will be passed to expression correlation assessment.
      For TFs not covered by interaction database, \emph{GRNBoost2}\cite{grnboost2}-like algorithm is initiated to predict potential regulatory target genes.
      The prediction importance threshold can either be set manually or automatically based on recorded interactions as:

      \centerline{$Threshold = min(GA(M_1, t)\, \uplus \, GA(M_2, t), \frac{1}{g})$}

      \noindent Here $GA()$ denotes \emph{GRNBoost2}\cite{grnboost2}-like algorithm;
      $M_1$ denotes concatenated class 1 samples GEM with genes passed filters;
      $M_2$ denotes concatenated class 2 samples GEM with genes passed filters;
      $t$ denotes the TF having largest amount of recorded interaction in dataset;
      $g$ denotes total amount of unique genes in all samples passed filters.

      After potential GRP gene pairs are obtained, an expression correlation filter is used to exclude gene pairs with low covariance.
      To assess expression correlation, AGEAS applies Pearson's Correlation coefficient\cite{pcc_2012} (PCC), one of the widely adopted methods.\cite{cid_2019}
      With default setting, gene pairs can reach absolute correlation coefficient of 0.2 and p-value lower than 0.05 are included in meta-GRN as validated GRP.

    \subsubsection*{Reconstruct psGRNs}
      The comprehensive GEMs are divided into sample subsets, and sliding window algorithm (SWA) is used to build pseudo-samples.
      With SWA, we can gain GEMs of \emph{i-th} pseudo-sample through:

      \centerline{$SWA(i) = \left\{{x_j}_{j = i * p}^{j + k}\right\}, j + k < l$}

      \noindent Here $l$ denotes number of samples in a comprehensive GEM which can be expressed as $\left\{{x_j}_{j = 0}^{l}\right\}$; $k$ denotes window size; $p$ denotes padding stride.

      If sample amount is considerably low or imbalanced, customized window size and padding stride can be applied to generate sufficient pseudo-samples for later classifier training and assessment processes.
      Utilizing meta-GRN, psGRNs are reconstructed with GEMs of pseudo-sample.
      Each GRP gene pair in meta-GRN is formed with expression data in pseudo-sample and is filtered by the same PCC filter in meta-GRN reconstruction process.

      After all pseudo-samples have been used to reconstruct psGRN, every psGRN can be represented as a matrix comprised of GRPs' PCC values.
      The order of GRPs in each psGRN is also unified through adding GRPs included in other psGRNs with 0.0 PCC value.

  \subsection*{Step 2: Classification model selection}
    \label{step2}
    The primary goal of AGEAS is to gain insights from multiple models as divergent as possible instead of developing optimized classification model for psGRN classification.
    Thus, this step is designed to select model configurations capable of predicting psGRN's masked sample class label.
    Considering the limited computational resources, the portion of less efficient model configurations shall be pruned at the cost of potentially losing insight in sample class difference.
    Therefore, we apply a simple Hyperband\cite{hyperband}-based algorithm \ref{alg:one} which performs a grid search for well-performing classification models with varying model training resources and pruning aggressiveness.


    %% This declares a command \Comment
    %% The argument will be surrounded by /* ... */
    \SetKwComment{Comment}{/* }{ */}
    \SetKwInput{kwInput}{Input}
    \SetKwInput{kwReturn}{Return}
    \RestyleAlgo{ruled}
    \begin{algorithm}
    \caption{Model selection algorithm}
    \label{alg:one}
    \kwInput{$R, C, I$ (default $I = 3$), $\alpha_{max}$(default $\alpha_{max} = 0.9$),  $k_{min}$(default $k_{min} = 0.5$)}
    $\alpha_{low} = \frac{1}{(2 ^ {I} - 1)}$\;
    % $X \gets x$\;
    \For{$i \in \{0, 1, ..., I-1\}$}{
      \eIf{$i == I-1$}{
        $\alpha = \alpha_{max}$\;
        % $X \gets X \times X$\;
        % $N \gets \frac{N}{2}$
        % \Comment*[r]{This is a comment}
      }{
        $alpha = 2^i \alpha_{low}$\;
      }
      $r = \alpha R$\;
      $P = \emptyset $\;
      \For{$c \in C$}{
        $p = run\_then\_evaluate(c, r, R)$\;
        Append $p$ to $P$\;
      }
      % \If{$N$ is odd}{
      %   }

      $k = max(1 - \alpha, k_{min}$)\;
  	  $C = top\_configs(C, P, k$)\;
    }
    \kwReturn{$C$}
    \end{algorithm}

    The model selection algorithm requires five inputs
    (1)$R$, the maximum amount of training resource, equivalent to all available psGRNs
    (2)$C$, the total set of provided classification model configurations
    (3)$I$, the number of iterations for model pruning (default set as 3)
    (4)$\alpha_{max}$, the maximum portion of $R$ can be fed to single model (default set as 0.95)
    (5)$k_{min}$, the minimum portion of remaining model configurations will be kept by single pruning iteration (default set as 0.5).
    Furthermore, two required functions need to be defined based on input model configurations:
    \begin{itemize}
    \setlength\itemsep{0em}
    \item \textbf{$run\_then\_evaluate(c, r, R)$}: trains classification model initialized using configuration $c$ for the allocated resource $r$, then returns prediction accuracy (ACC), the area under a receiver operating characteristic curve (AUROC)\cite{hanley_mcneil_1982} score, and total cross-entropy loss ($L_{CE}$) calculated through predicting sample class for all psGRNs $R$.

    \item \textbf{$top\_configs(C, P, k)$}: takes a set of model configurations $C$ with associated evaluation results $P$ and returns configurations with ACC, AUROC score, or $L_{CE}$ reaching top $k$ portion.
    \end{itemize}

    \noindent By default, AGEAS initializes with 128 model configurations utilizing 9 integrated classification algorithms listed in Table \ref{models}.

    \begin{table}[ht]
      \centering
      \begin{tabular}{|l|l|l|l|l|}
      \specialrule{.2em}{.1em}{.1em}
      \textbf{Algorithm} & \textbf{$\# \lambda$} & \textbf{Categorical} & \textbf{Continuous} & \textbf{$\#$ Configs}\\
      \specialrule{.2em}{.1em}{.1em}
      \multicolumn{5}{|l|}{\emph{Implemented with Pytorch}\cite{pytorch}} \\
      \hline
      Transformer & \textbf{14} & 4 & 10 & 32 \\
      \hline
      1D Convolutional Neural Network (1D-CNN) & \textbf{10} & 2 & 8 & 32 \\
      \hline
      Hybrid Convolutional Neural Network (Hybrid-CNN) & \textbf{10} & 2 & 8 & 32 \\
      \hline
      Gated Recurrent Unit (GRU) & \textbf{10} & 4 & 6 & 4 \\
      \hline
      Long Short-Term Memory (LSTM) & \textbf{11} & 4 & 7 & 4 \\
      \hline
      Standard Recurrent Neural Network (RNN) & \textbf{11} & 5 & 6 & 4 \\
      \specialrule{.2em}{.1em}{.1em}
      \multicolumn{5}{|l|}{\emph{Implemented with XGBoost}\cite{chen2016xgboost}} \\
      \hline
      Gradient Boosted Decision Trees (GBDT) & \textbf{18} & 5(1) & 13(4) & 16 \\
      \specialrule{.2em}{.1em}{.1em}
      \multicolumn{5}{|l|}{\emph{Implemented with scikit-learn}\cite{scikit-learn}} \\
      \hline
      Random Forests (RF) & \textbf{14} & 5(1) & 9(1) & 2 \\
      \hline
      Support Vector Machine (SVM) & \textbf{7} & 4 & 3(3) & 2 \\
      \specialrule{.2em}{.1em}{.1em}
      \end{tabular}
      \caption{
        \label{models}
        Classification model algorithms integrated in AGEAS with corresponding numbers of hyperparameter and preset model configurations.
        Categorical hyperparameters and continuous numerical hyperparameters are clarified beside total number of hyperparameters ($\# \lambda$).
        Conditional hyperparameters required for selected other hyperparameters are shown in brackets.
        $\#$ Configs indicates the default total number of configurations.
      }
    \end{table}

    The general architecture designs of 1D-CNN and Hybrid-CNN are implemented based on 1D-CNN and 2D-Hybrid-CNN in a recent cancer type prediction study.\cite{mostavi_chiu_huang_chen_2020}
    However, taking one convolution layer and adjacent max-pooling layer as a layer set, we implemented both CNN models with flexibility on the number of layer set, which is fixed in the original paper.
    An example of 1D-CNN with two convolution layer sets is illustrated in Figure \ref{1dCNN}.

    For transformer models, the embedding layer and positional encoding layer designed for input data tokenization in standard architecture \cite{transformer} are replaced with a single linear layer in AGEAS, considering psGRNs are already represented as numerical data matrix while GRPs should barely have positional relationships in the matrix.

  \subsection*{Step 3: Feature selection}
    \label{step3}
    With selected well-performing models, AGEAS already can start repetition of model training and interpretation in \hyperref[step4]{Step 4} to extract key GRPs.
    However, the few uninformative GRPs in training psGRNs could be pruned in advance to save computational resource.
    Furthermore, to prevent classification models from focusing on a small group of GRPs regardless of training psGRNs, GRPs draw excessive attention shall be separated from psGRNs to improve extraction comprehensiveness.
    Thus, in this step, AGEAS iteratively trains classification models with dynamic $\alpha_{max}$ portion of psGRNs and obtains feature importance scores as described in \hyperref[features_importances]{subsection} below to exclude GRPs either scored considerably high or relatively low.

    More specifically, at each iteration, GRPs with z-scores ranked at the bottom $b$ portion (default set as 0.1) are discarded.
    Also, an $i$-th ranked GRP will be separated from psGRNs and passed to \hyperref[step5]{Step 5} directly if it has a z-score fulfilling the condition:

    \centerline{
      $Z_{score}^{i} \ge max(Z_{score}^{thread}, \frac{Z_{score}^{i-1}}{3}, 3 \cdot IQR)$
    }

    \noindent The $Z_{score}^{thread}$ denotes input z-score threshold (default set as 3.0), and $IQR$ stands for interquartile range calculated at the beginning of each iteration.\
    With this criterion, AGEAS ensures only GRPs draw significantly more attention be selected by each iteration, despite different data distribution of z-score scaled importance values.

    By default, AGEAS iterates this feature selection step for three times.

    \subsubsection*{Feature importance estimation}
      \label{features_importances}
      AGEAS applies concept of The Shapley value\cite{roth_1988} for estimating the importance of each input feature, equivalent to GRP of input psGRN, in any kind of classification model when making predictions.
      Specific Shapley value calculation or approximation methods are implemented with \emph{SHAP}\cite{lundberg2017unified} and applied to different algorithms as shown in Table \ref{shap}.
      We utilize \emph{softmax} function to normalize feature importance scores and define the normalization function as:

      \centerline{$T(X) = softmax(\left\{ F(x) \right\}, x \in X)$}

      \noindent Here $X$ denotes the total set of all features, and $F(x)$ denotes importance score estimation method.\
      If the feature importance can be approached with internalized method $f(x)$, $F(x)$ is set as:

      \centerline{$F(x) = \frac{f(x)}{\sum_{x' \in X} f(x')}$}

      \noindent Otherwise, $F(x)$ is defined utilizing correctly classified input samples $S$, equivalent to psGRNs, with Shapley values $\phi$ of feature $x$ when predicting sample $s$ as class $c_1$ or class $c_2$:\

      \centerline{$F(x) = \sum_{s \in S}\frac{\left|\phi_{c_1,s}^{x}\right| + \left|\phi_{c_2,s}^{x}\right|}{2}$}

      \noindent After all selected classification models $M$ have been interpreted, we can integrate the feature importance matrices weighted by the corresponding models' $L_{CE}$ to one matrix $A$ as:\

      \centerline{
        $ A = \left\{ \sum_{m \in M}(1 - L_{CE}^{m})T_{m}(x) \right\}, x \in X $
      }

      \noindent Then, generalized importance values are obtained through z-score calculation and later sorted descendingly:\

      \centerline{
        $Z_{score} = \left\{\frac{a - \bar{A}}{\sigma_A}\right\}, a \in A$
      }

      \begin{table}[ht]
        \centering
        \begin{tabular}{|l|l|}
        \specialrule{.2em}{.1em}{.1em}
        \textbf{Algorithm} & \textbf{SHAP\cite{lundberg2017unified} Method}\\
        \hline
        Transformer & Gradient Explainer \\
        \hline
        1D-CNN & Deep Explainer \\
        \hline
        Hybrid-CNN & Deep Explainer \\
        \hline
        GRU & Gradient Explainer \\
        \hline
        LSTM & Gradient Explainer \\
        \hline
        RNN & Gradient Explainer \\
        \hline
        GBDT\textbf{*} & Tree Explainer \\
        \hline
        RF & Tree Explainer \\
        \hline
        SVM\textbf{*} & Linear Explainer / Kernel Explainer\\
        \specialrule{.2em}{.1em}{.1em}
        \end{tabular}
        \caption{
          \label{shap}
          Classifier algorithms with applicable Shapley value approximation methods.
          Algorithms marked with \textbf{*} have internalized feature importance estimation methods which will be applied with higher priority than Shapley value based methods.
          GBDT implemented with \emph{XGBoost}\cite{chen2016xgboost} can have feature importance approximated by the average weight gain at each split involving the feature.
          Linear SVM implemented with \emph{scikit-learn}\cite{scikit-learn} can have importance scores estimated with feature coefficients or using Linear Explainer.
          However, the feature coefficient estimation should be inappropriate for SVM with kernel function, While feature importance scores should be approximated with Kernel Explainer only.
        }
      \end{table}

  \subsection*{Step 4: Top GRP extraction}
    \label{step4}
    To extract GRPs capable of effectively defining sample class differences, AGEAS iteratively initializes classification models with configurations kept from \hyperref[step2]{Step 2}, trains them with randomly selected $\alpha_{max}$ portion of rescaled psGRNs from \hyperref[step3]{Step 3}, and interprets every models' accurate predictions with methods mentioned in \hyperref[features_importances]{subsection}.
    At each iteration, AGEAS receives a feature importance matrix $A$ re-scaled by z-score and adds up each GRP's score accordingly from matrix $A'$ kept from previous iteration if available.
    Then, top $a$, by default 100, ranked GRPs, whose scores are also greater than $0.0$, are compared with GRPs extracted from $A'$ by the same setting.
    If changing portion of GRPs for top scored GRP sets from $A$ and $A'$ is less than $d$, by default 0.05, AGEAS will consider the GRP extraction result of current iteration consistent with previous iteration.
    Extraction loop will terminate if AGEAS encounters $n$ (default set as 3) continuous consistent result or runs out of preset iteration number (default set as 10).
    All feature importance scores in matrix $A$ from last iteration are divided by the total iteration number processed, and top $a$ ranked GRPs are considered as important GRPs for sample class differentiation.

  \subsection*{Step 5: Key network reconstruction}
    \label{step5}
    Analogizing every GRP previously extracted in \hyperref[step4]{Step 4} or separated as outlier in \hyperref[step3]{Step 3} as a directional edge connecting a TF vertex with a targetome vertex, AGEAS attempts to reconstruct a GRN graph representing regulatory differences between query sample classes.
    Since there is no guarantee on whether all extracted GRP edges can be connected or not, some regulatory relationships between gene vertices could be absent.
    Hence, AGEAS utilizes meta-GRN gained from \hyperref[step1]{Step 1} to find GRPs, which can further elucidate the regulatory relationships, and to use those GRPs as supportive edges.

    For a limited iteration time (default set as  1), AGEAS exhaustively searches meta-GRN for TFs which can directly interact with any gene already covered in extracted GRPs and add the returned TFs as new vertices.
    Next, any GRP in meta-GRN capable of connecting two distinct vertices will be included as supportive GRP in extraction result.
    After expansion procedure above, the extracted GRN graph represents key genetic regulatory differences between input sample classes.

\section*{Results}
  \label{res}
  To assess predictive power of utilizing extraction result of AGEAS, we primarily applied AGEAS to study somatic cell reprogramming to induced pluripotent stem cells (iPSCs) in mice, \cite{yamanaka_2006} a milestone discovery of cell plasticity. \cite{cell_repro_review}
  With $\sigma$ thread set to 2.0 for constraining total GRP amount in psGRNs, we use public scRNA-seq based GEMs of embryonic stem cells (ESCs) and mouse embryonic fibroblasts (MEFs) shown in Table \ref{geo_table} as inputs.
  The extraction result can be summarized as TF regulons, and TFs with highest regulatory degree in extracted GRN atlas are marked as top TFs (Figure \ref{result_figs}a).

  Among top TFs, Pou5f1 (Oct4) associates with GRPs extracted as important differential regulatory element between MEF and ESC more than any other TFs, while forming largest regulon and having expression pattern significantly favoring ESC.
  Thus, we infer Pou5f1 is the key regulatory element associated with ESC identity.
  Extensive studies already addressed Pou5f1's important role in pluripotency maintenance. \cite{niwa_2007, oct4_1, oct4_2}
  Moreover, to implement cell reprogramming of iPSC, Nanog and Sox2 are noteworthy, since they closely interact with Pou5f1 in extracted network, have high regulatory degrees, and are differently expressed in ESC.
  Multiple previous studies confirmed that all of Oct4, Nanog, and Sox2 are playing important roles to induce cell conversion from somatic cell to iPSC. \cite{yamanaka_2006, ips7f, ipsOK, oct4_nanog_sox2_lin28, oct4_nanog_sox2}
  Cebpb, Hdac2, and Smc3, extracted as top TFs directly affecting Pou5f1 expression, were also reported with relevant functions. \cite{cebpb_ipsc, hdac2_ipsc, rad21_smc3_esc}

  Therefore, we infer that the investigation of TFs as common regulatory sources in important differential GRPs and corresponding regulons can effectively reflect differences between cell types or cell states.
  % \subsection*{In silico evaluations}
  To further address AGEAS's applicability and limitation, we applied AGEAS to three other scRNA-seq based studies in distinct research areas.
  With default settings described in \hyperref[method]{Method} section above, we evaluated AGEAS's performance in following scenarios with public dataset (Table \ref{geo_table}):

  \begin{itemize}
      \setlength\itemsep{0em}
      \item {\textbf{Dopaminergic neuron differentiation}}:

        We address the applicability of AGEAS on cell subtype differentiation problems by analyzing the difference between human iPSC-derived radial glial / neuronal co-culture as neural progenitors\cite{ASCL1_dopaminergic_neuron_2021} and tyrosine hydroxylase (TH) expressing purified dopaminergic neurons (DANs).

        From extracted TFs (Figure \ref{result_figs}b), we hypothesized that ISL1, ELF3, and PBX3 play important roles in DAN differentiation due to their high regulatory degree in extracted important GRPs.
        Previously, ISL1 was determined essential for differentiation of prethalamic DANs. \cite{isl1_da}
        ELF3 was found to be the neuronal precursor cell marker associated with neural stem cell development. \cite{ELF3_precursor_marker}
        This is consistent with ELF3's high expression level in neuronal co-culture samples.
        Moreover, a DAN development study reported that PBX3 is required for correct differentiation of neuroblast, production of radial glial cell, into midbrain DAN and survival of midbrain DAN. \cite{pbx3_dan}

        Through stratifying all extracted GRPs, we only found that HDAC2 and RBPJ can potentially regulate all of ISL1, ELF3, and PBX3.
        Few reports demonstrated the association of HDAC2 with neurogenesis of radial glial cells. \cite{hdac2_dan_1, hdac2_dan_2}
        Nevertheless, HDAC2's role in DAN differentiation is yet to be clarified.
        RBPJ was reported to be essential for DAN survival and to affect DAN development by regulating ASCL1, a essential factor in neurogenesis. \cite{rbpj_1, rbpj_2}
        The original research where we retrieved scRNA-seq data from also demonstrated that ASCL1 is a necessary factor to regulate DA neurotransmitter selection. \cite{ASCL1_dopaminergic_neuron_2021}
        Although AGEAS extracted ASCL1 as one of the top TFs, no direct regulatory relationship between RBPJ and ASCL1 was identified.
        % PHOX2B, and ELF3 are regulating essential functions differentiating DANs from neural progenitors referring to their correspond regulon sizes and differential expression profiles.
        % Both ISL1 and PHOX2B were found related with DA phenotype in previous studies:
        % ISL1 is
        % and PHOX2B is connected with TH expression influencing DAN development and improving DA activity. \cite{phox2_caudal_da, phox2_rat_da}

        % EBF1, which forms the largest regulon with extracted GRPs, was determined as key neuronal differentiation regulator in the ventral telencephalon, nature source of neuronal and glial cells. \cite{ebf1_striatum, ebf1_cell_diff}
        % Furthermore, EBF1 plays important role in mesodiencephalic DAN migration. \cite{ebf1_migration}


      \item {\textbf{Postnatal cardiomyocyte maturation}}:

        To address the performance of AGEAS to analyze cell physiological development, scRNA-seq data of cardiomyocytes (CMs) in postnatal day 7 mice (P7) and day 28 mice (P28) are used as inputs for extracting key genetic regulatory elements in postnatal CM maturation.

        As indicated by important differential GRPs, we first investigated Sp1 and Esr1 ($ER\alpha$) that also have top regulatory degrees and considerably differential expression profiles (Figure \ref{result_figs}c).
        Previous studies demonstrated that Sp1 promotes CM hypertrophy \cite{sp1_hypertrophy} and maturation of electrophysiology, $Ca^{2+}$ handling. \cite{Sp1_electrophysiologt, CM_mature}
        % trans-activation of the SERCA2 promoter
        Esr1 was found modulating myocardial development for postnatal cardiac growth.\cite{esr1_cm, esr1_cm_growth}

        Beside the inter-regulation between Sp1 and Esr1, three other top TFs (Srf, Cebpb, Rest) and two TFs forming relatively small regulons (Gata4, Prox1), are expected to interact with both Sp1 and Esr1 based on extraction network.
        All five TFs were found to have CM maturation related functions. (Table \ref{interact_TFs})

        \begin{table}[ht]
            \centering
            \begin{tabular}{|l|l|}
              \hline
              \textbf{Gene} & \textbf{Function}  \\
              \hline
              Srf & Regulate sarcomere genes and broadly impact almost every aspect of CM maturation\cite{CM_mature} \\
              \hline
              Cebpb & Repress CM proliferation and hypertrophy\cite{cebpb_1, cebpb_2}\\
              \hline
              Rest (Nrsf) & Repress CM hypertrophy; Prevent dilated cardiomyopathy\cite{rest_1}\\
              \hline
              Prox1 & Repress CM hypertrophy; Prevent dilated cardiomyopathy\cite{prox1_CM} \\
              \hline
              Gata4 & Promote CM proliferation and hypertrophy\cite{CM_posnatal_mature} \\
              \hline
            \end{tabular}
            \caption{
              \label{interact_TFs}
              TFs interacting with both Sp1 and Esr1 order by regulatory degree in extracted GRNs.
            }
        \end{table}

        % Other top TFs expected to directly influence expression of either Sp1 or Esr1 are:
        % \begin{table}[ht]
        %     \centering
        %     \begin{tabular}{|l|l|}
        %       \hline
        %       \textbf{Gene} & \textbf{Function}  \\
        %       \hline
        %       Nfe2l2 &  mitochondrial and cardiac oxidative stress, cardiac fibrosis (10.1089/ars.2016.6831) \\
        %       \hline
        %       Nfib & 1 \\
        %       \hline
        %       Cbfb & 1 \\
        %       \hline
        %       Fosl2 & 1 \\
        %       \hline
        %       Nkx2-5 & \cite{cebpb_1, nkx25, CM_posnatal_mature} \\
        %       \hline
        %       Rxra & 1 \\
        %       \hline
        %       Mef2a & 1 \\
        %       \hline
        %       Yap1 & \cite{CM_mature, yap1, CM_posnatal_mature} \\
        %       \hline
        %       Thra & 1 \\
        %       \hline
        %     \end{tabular}
        %     \caption{
        %       \label{interact_TFs}
        %       Extracted top TFs interacting with either Sp1 or Esr1.
        %     }
        % \end{table}

      \item {\textbf{Hepatic stellate cell activation in $CCl_4$ induced liver fibrosis}}:

        Here we address AGEAS's applicability on cell pathological development studies through analyzing hepatic stellate cells (HSCs) and portal fibroblasts (PFs) simulating activated HSCs in mice liver administrated with carbon tetrachloride ($CCl_4$) for six weeks.

        Based on extracted important GRPs, we identified Mef2c, Jun, and JunD as potential representatives of key regulatory elements in HSC activation (Figure \ref{result_figs}d).
        As the TF extracted with most important GRPs, Mef2c has been well-documented as the key regulator in HSC activation. \cite{mef2c_1, mef2c_2, mef2c_3}
        Extensive studies also found Jun and JunD, functional components of the AP1 TF complex, are essential for fibrosis-related HSC activation and profibrogenic process. \cite{ap1_hsc_1, ap1_hsc_2, ap1_hsc_3, ap1_hsc_4}

        Analyzing TF regulons of Mef2c, Jun, and JunD, we believe that Mef2c is targetome of JunD and the inter-regulatory element of Jun.
        While no regulatory relationship between JunD and Jun determined, three extracted top TFs (Lhx2, Nfib, JunB) can potentially affect expressions of both JunD and Jun.
        JunB, another functional component of AP1, is also reported to  affect HSC activation \cite{ap1_hsc_5}.
        Furthermore, Lhx2 was found indispensable for quiescent HSCs. \cite{lhx2_fibro, Lhx2_hsc_1}
        Although little literature had addressed Nfib's role in HSC, Nfib was reported to be the anti-apoptotic gene for cell proliferation during $CCl_4$ induced liver damage. \cite{Nfib_sadly}

    \end{itemize}

  % \subsection*{In vitro validation}
  %   To demonstrate empirically the predictive capabilities of utilizing AGEAS extraction result, we further investigated the extraction result of $CCl_4$ induced liver fibrosis study setting scenario of identifying theraputic targets inhibiting pathological fibrogenic process in liver.
  %   Through analyzing extracted TF regulons, we further confirmed Mef2c's role as key targetome noticing direct regulational relationships with HSC activation marker Acta2\cite{acta2_hsc_act} and fibrotic collagen Col1a1 (Figure \ref{regulons_figs}d).
  %   Moreover, since only Lhx2 has all related GRPs mainly being functional in HSC (Figure \ref{regulons_figs}a) and significantly differential expression in HSC (Figure \ref{result_figs}d), we decided to proceed manipulation of Lhx2 expression.
  %   Previous study already proved knockdown of Lhx2 promotes liver fibrosis in mice.\cite{lhx2_fibro}
  %   Thus, we suppose overexpression of Lhx2 could assist to maintain quiescent of HSCs through Jun/Jund-Mef2c dependent pathways.
  %
  %   To verify our hypothesis, we firstly validated the association between Lhx2 and HSC activation with $in$ $vitro$ experiments.
  %   In mice primary HSC (mHSC) and human HSC cell lines (LX2), we detected the expression of Lhx2/LHX2 with the activation of HSC.
  %   The results showed that the expression of Lhx2 was significantly decreased with the activation of HSC (Figure \ref{hsc_activation_mRNA}), consistent with our previous analysis result.
  %   To further clarify the role of LHX2 in human cell, we knocked down LHX2 by small interfering RNA (siRNA) in LX2 cells.
  %   The results showed that knockdown of LHX2 significantly promoted the expression of fibrosis-related genes in Figure \ref{hsc_activation_mRNA}d, compatible with previous knowledge in mice.
  %
  %   Subsequently, we overexpressed Lhx2/LHX2 in mHSCs and LX2 cells.
  %   The results showed that overexpression of LHX2 directly affected the cell morphology of human HSCs and significantly inhibited the expression of fibrogenic genes (Figure \ref{LHX2_LX2_morphology} and \ref{LHX2_OEs}a).
  %   In addition, observation in mHSCs further confirmed that Lhx2 could inhibit the expression of fibrosis-related genes (Figure \ref{LHX2_OEs}b).
  %   In order to comprehensively understand the gene expression profile of HSC regulated by LHX2, RNA-seq results showed that the differential genes were mainly related to extracellular matrix structure and assembly (Figure \ref{LHX2_GO}), which was consistent with fibrosis-related phenotype.
  %
  %   Thus, systematic and comprehensive $in$ $vitro$ verification showed that Lhx2/LHX2 is indeed capable to regulate fibrogenic gene expressions and inhibit HSC activation as we concluded from AGEAS extraction output.
  %   However, the preliminarily identified regulatory pathways for Lhx2 to inhibit HSC activation is yet to be addressed.
  %   % In vivo experiment will take longer time...

\section*{Conclusion}
  In summary, AGEAS shows robustness in extracting key regulatory elements in several biological studies.
  Key TFs inferred by important differential GRPs and regulon analysis in all four $in$ $silico$ cases were confirmed informative to reveal main functionalities responsible for biological process of interest.
  Furthermore, the direct extractions of ISL1 in DAN differentiation and JunD in HSC activation demonstrated that AGEAS is neither focusing solely on the regulatory degrees in reconstructed meta-GRN nor the expression differences between sample classes.
  % Also, $in$ $vitro$ experiments indicate great potential in AGEAS to promote method development in phenotypic manipulation.
  % While TF regulons reconstructed with all extracted GRPs are having sizes proportional to meta-GRN... GO enrichment?
  % However, whether extracted regulatory elements covering all functional aspects of studying biological phenomena, admittedly, remains undetermined due to lack of validation for GRPs' activities.
  Overall, we anticipate that AGEAS is useful in providing insightful regulatory information and in promoting pioneer researches that reveal complicated mechanism behind biological phenomena.

\section*{Discussion}
  \label{disc}
  Considering the rapid development of both sequencing technology and machine learning algorithms, we implemented AGEAS with modular design.
  In the data preprocessing part, the meta-GRN reconstruction section can be replaced with other methods to further increase fidelity of regulatory relationships or integrating more information, such as cis-regulatory elements (CREs) accessibility, which can be obtained from Assay for Transposase-Accessible Chromatin Sequencing (ATAC-seq) data.
  Moreover, CRE accessibility information combining with motif enrichment analysis is proved capable of identifying potential regulatory pathways to replace ChIP-seq based dataset utilized in the current study. \cite{scREMOTE}
  Refining meta-GRN reconstruction method in consideration of having input data from simultaneous scRNA-seq and single cell ATAC-seq (scATAC-seq), such as SHARE-seq published in late 2020, \cite{share_seq} can potentially further improve applicability of AGEAS; however, accessibility of corresponding datasets will be essential for performance assessment and method development.
  Similarly, initial set of classification models is also subject to change in response to future computational studies.

  Also, it is important to note that AGEAS can be applied with GEMs normalized with different methods on user's choice, as well as derived from other sequencing technologies, such as bulk RNA-seq or spatial sequencing in revealing tissue differences.
  Replacing transcriptiomic data with proteomic data may also be achievable in the analysis of protein interaction networks.
  However, larger scale of applicability tests are necessary.

  Due to the stochastic nature of AGEAS, extraction results of AGEAS are not expected to stay identical for repeated applications even with exactly the same inputs.
  The main reason underlying the inconsistency is how AGEAS attempts to obtain top performing models for later prediction interpretations.
  Repeated random subsets of psGRNs used in model training and selection are expected to vary for each application, hence results in different performances of classification models and ranking results.
  As a result, the importance scores of GRPs might differ and slightly affect top TFs extracted later.

  Two possible workarounds of the inconsistent issue would be:
  (i) manually set random seed for every process influencing subset selection.
  (ii) set multiple extractor unit and combine extraction results.
  In the present study, we found that n = 4 units yield relatively stable results, but this value could vary based on application scenario.

  % Although Lhx2 extracted by AGEAS-based analysis indeed shows capability to inhibit HSC activation during liver fibrosis as predicted, the corresponding inference of functional pathways still need to be proved.
  % Furthermore, $in$ $vivo$ experiments are also necessary to evaluate functionality of Lhx2 and associated pathways for higher fidelity.

\section*{Acknowledgement}
  We sincerely thank Cher Li and Jie Wang for inspiration; Augustine(Xiaoran) Yuan and Elizabeth Bania for reviewing and support.

\bibliography{reference}

% For data citations of datasets uploaded to e.g. \emph{figshare}, please use the \verb|howpublished| option in the bib entry to specify the platform and the link, as in the \verb|Hao:gidmaps:2014| example in the sample bibliography file.

\section*{Author contributions statement}
  \textbf{J.Y.}: Methodology, Software, Writing- Original draft preparation, Project administration
  \textbf{M.N.}: Methodology, Software
  \textbf{J.T.}: Methodology, Writing- Original draft preparation, Project administration
% Writing - Review \& Editing,

\section*{Additional information}
  All scRNA-seq datasets are retrieved from Gene Expression Omnibus(GEO) as described in the Table below.

  \begin{table}[ht]
    \centering
    \begin{tabular}{|l|l|}
    \specialrule{.2em}{.1em}{.1em}
    \textbf{Sample Class} & \textbf{Accession number}\\

    \specialrule{.2em}{.1em}{.1em}
    \multicolumn{2}{|l|}{GSE103221} \\
    \hline
    MEF & GSM3629847 \\
    \hline
    ESC & GSM3629848 \\

    \specialrule{.2em}{.1em}{.1em}
    \multicolumn{2}{|l|}{GSE137720 } \\
    \hline
    $CCl_4$ a6w hsc & GSM4085625 \\
    \hline
    $CCl_4$ a6w pf & GSM4085627 \\
    % \hline
    % healthy hsc & GSM4085623 \\
    % \hline
    % healthy pf & GSM4085626 \\

    \specialrule{.2em}{.1em}{.1em}
    \multicolumn{2}{|l|}{GSE185275 } \\
    \hline
    glial/neuronal co-culture & GSM5609927 \\
    \hline
    purified DANs & GSM5609930 \\

    \specialrule{.2em}{.1em}{.1em}
    \multicolumn{2}{|l|}{GSE156482 } \\
    % \hline
    % p4 cm & GSM4732219 \\
    \hline
    p7 CM & GSM4732221 \\
    \hline
    p28 CM & GSM4732225 \\
    % \hline
    % p28 ncm & GSM4732226 \\

    \specialrule{.2em}{.1em}{.1em}
    \end{tabular}
    \caption{
      \label{geo_table}
    }
  \end{table}

\begin{figure}[ht]
  \centering
  % \includegraphics[width=0.8\linewidth, height=18cm, keepaspectratio,]{../images/summary_trans.png}
  \includegraphics[width=0.8\linewidth, keepaspectratio,]{../images/summary_trans.png}
  \caption{
    The overall workflow of AGEAS:
    \textbf{\emph{(1)}} Reconstruct meta-level GRN (meta-GRN) with expression data of all samples.
    \textbf{\emph{(2)}} Build pseudo-samples with sliding window algorithm and reconstruct GRNs with GRPs identified in meta-GRN accordingly.
    \textbf{\emph{(3)}} Select best performing classifiers in predicting class labels of pseudo-sample GRNs (psGRNs).
    \textbf{\emph{(4)}} Interpret how top models make classifications and gradually exclude GRPs with low weights or outlier-level high weights.
    \textbf{\emph{(5)}} Repeatedly train classifiers with different set of psGRNs as training data to extract GRPs frequently ranked as top important features for decision.
    \textbf{\emph{(6)}} Reconstruct GRNs with extracted GRPs and GRPs excluded as significant outliers.
  }
  \label{workflow}
\end{figure}

\begin{figure}[ht]
  \centering
  % \includegraphics[width=0.8\linewidth, height=12cm, keepaspectratio,]{../images/data_preprocess_trans.png}
  \includegraphics[width=0.8\linewidth, keepaspectratio,]{../images/data_preprocess_trans.png}
  \caption{
    Workflow to reconstruct meta-GRN and psGRNs.
    \textbf{\emph{(1)}} Filter genes from GEMs with log2FC filter(optional), MWW filter, and $\sigma$ filter.
    \textbf{\emph{(2)}} Find candidate GRP gene pairs from either interaction database or predictions made by \emph{GRNBoost2}\cite{grnboost2}-like algorithm.
    \textbf{\emph{(3)}} Filter candidate GRP gene pairs with PCC filter and reconstruct meta-GRN with validated GRPs.
    \textbf{\emph{(4)}} Generate pseudo-samples with SWA.
    \textbf{\emph{(5)}} Utilize GRPs in meta-GRN as generic guidance to form candidate GRPs for pseudo-samples.
    \textbf{\emph{(6)}} Filter every candidate GRP for all pseudo-samples with PCC filter and reconstruct psGRNs with validated GRPs.
  }
  \label{data_preprocess}
\end{figure}

\begin{figure}[ht]
  \centering
  \includegraphics[width=0.8\linewidth]{../images/nn.png}
  \caption{1D-CNN with 2 convolution layer set.}
  \label{1dCNN}
\end{figure}

\begin{figure}
  \begin{subfigure}{0.49\linewidth}
    \centering
    \includegraphics[width=\linewidth, keepaspectratio,]{../images/top_TFs/MEF_ESC.png}
    \caption{}
  \end{subfigure}
  \begin{subfigure}{0.49\linewidth}
    \centering
    \includegraphics[width=\linewidth, keepaspectratio,]{../images/top_TFs/N_cocul.png}
    \caption{}
  \end{subfigure}

  \begin{subfigure}{0.49\linewidth}
    \centering
    \includegraphics[width=\linewidth, keepaspectratio,]{../images/top_TFs/CMp7d_CMp28d.png}
    \caption{}
  \end{subfigure}
  \begin{subfigure}{0.49\linewidth}
    \centering
    \includegraphics[width=\linewidth, keepaspectratio,]{../images/top_TFs/HSCp6w_PFp6w.png}
    \caption{}
  \end{subfigure}

  \caption{
    Top 15 TFs ranked by regulatory degrees in all GRPs extracted by AGEAS.
    % Heatmap at right indicates relative $log_2$ gene expression levels in binary sample classes.
    \textbf{\emph{(a)}} Mouse embryonic fibroblast vs. Embryonic stem cell
    \textbf{\emph{(b)}} Purified dopaminergic neuron vs. Radial glial/neuronal co-culture
    \textbf{\emph{(c)}} 7 days postnatal cardiomyocyte vs. 28 days postnatal cardiomyocyte
    \textbf{\emph{(d)}} Hepatic stellate cell vs. Portal fibroblast (both after 6 weeks of CCl4 administration)
  }
  \label{result_figs}
\end{figure}

% \begin{figure}
%
%   \begin{subfigure}{0.49\linewidth}
%     \centering
%     \includegraphics[width=\linewidth, keepaspectratio,]{../images/lhx2_class.png}
%     \caption{Lhx2 regulon}
%   \end{subfigure}
%   \begin{subfigure}{0.49\linewidth}
%     \centering
%     \includegraphics[width=\linewidth, keepaspectratio,]{../images/junb_class.png}
%     \caption{Junb regulon}
%   \end{subfigure}
%
%   \begin{subfigure}{0.49\linewidth}
%     \centering
%     \includegraphics[width=\linewidth, keepaspectratio,]{../images/nfib_class.png}
%     \caption{Nfib regulon}
%   \end{subfigure}
%   \begin{subfigure}{0.49\linewidth}
%     \centering
%     \includegraphics[width=\linewidth, keepaspectratio,]{../images/mef2c_class.png}
%     \caption{Mef2c regulon}
%   \end{subfigure}
%
%   \caption{
%     Regulons of extracted top TFs interacting with both Jund and Jun and Mef2c.
%     Class 1 stands for HSC; class 2 stands for PF.
%   }
%   \label{regulons_figs}
% \end{figure}

% \begin{figure}
%   \begin{subfigure}{0.49\linewidth}
%     \centering
%     \includegraphics[width=\linewidth, keepaspectratio,]{../images/wet_lab/tao_a.png}
%     \caption{}
%   \end{subfigure}
%   \begin{subfigure}{0.49\linewidth}
%     \centering
%     \includegraphics[width=\linewidth, keepaspectratio,]{../images/wet_lab/tao_b.png}
%     \caption{}
%   \end{subfigure}
%
%   \begin{subfigure}{0.49\linewidth}
%     \centering
%     \includegraphics[width=\linewidth, keepaspectratio,]{../images/wet_lab/tao_c.png}
%     \caption{}
%   \end{subfigure}
%   \begin{subfigure}{0.49\linewidth}
%     \centering
%     \includegraphics[width=\linewidth, keepaspectratio,]{../images/wet_lab/tao_h.png}
%     \caption{}
%   \end{subfigure}
%   \caption{
%     Relative mRNA expression levels during HSC activation.
%     \textbf{\emph{(a)}} Fibrosis-related gene (Col1a1, Col3a1, Acta2) expression in mHSCs were increased with the extension of culture time, indicating HSC activation.
%     n = 3.
%     \textbf{\emph{(b)}} With the activation of mHSC, expression of Lhx2 decreased gradually.
%     n = 3.
%     \textbf{\emph{(c)}} In human HSC cell line LX2, TGF-$\beta$ can remarkably activate HSC and promote expressions of fibrosis-related genes. % Reference ?
%     Meanwhile, the expression of LHX2 decreased with the activation of HSC.
%     \textbf{\emph{(d)}} Knockdown of LHX2 by siRNA (si-LHX2) significantly increased the expression of fibrosis-related genes in LX2 cells compared with negative control (si-NC).
%     n = 4.
%   }
%   \label{hsc_activation_mRNA}
% \end{figure}
%
% \begin{figure}
%   \begin{subfigure}{0.49\linewidth}
%     \centering
%     \includegraphics[width=\linewidth, keepaspectratio,]{../images/wet_lab/tao_d_lx2_wt.png}
%     \caption{LX2-WT}
%   \end{subfigure}
%   \begin{subfigure}{0.49\linewidth}
%     \centering
%     \includegraphics[width=\linewidth, keepaspectratio,]{../images/wet_lab/tao_d_lx2_pcdh_oe.png}
%     \caption{LX2-PCDH-OE}
%   \end{subfigure}
%
%   \begin{subfigure}{0.49\linewidth}
%     \centering
%     \includegraphics[width=\linewidth, keepaspectratio,]{../images/wet_lab/tao_d_lx2_lhx2_oe.png}
%     \caption{LX2-LHX2-OE}
%   \end{subfigure}
%   \caption{
%     Overexpression of LHX2 significantly altered cell morphology in LX2 cell lines.
%     \textbf{\emph{(a)}} LX2 wild type.
%     \textbf{\emph{(b)}} LX2 overexpressing negative control vector.
%     \textbf{\emph{(c)}} LX2 overexpressing LHX2.
%   }
%   \label{LHX2_LX2_morphology}
% \end{figure}
%
% \begin{figure}
%   \begin{subfigure}{0.49\linewidth}
%     \centering
%     \includegraphics[width=\linewidth, keepaspectratio,]{../images/wet_lab/tao_e.png}
%     \caption{LX2}
%   \end{subfigure}
%   \begin{subfigure}{0.49\linewidth}
%     \centering
%     \includegraphics[width=\linewidth, keepaspectratio,]{../images/wet_lab/tao_f.png}
%     \caption{mHSC}
%   \end{subfigure}
%
%   \caption{
%     Overexpression of Lhx2/LHX2 significantly decreased the expression of fibrosis-related genes.
%     \textbf{\emph{(a)}} in LX2 cells. n = 3.
%     \textbf{\emph{(b)}} in mHSCs. n = 3.
%   }
%   \label{LHX2_OEs}
% \end{figure}
%
% \begin{figure}[ht]
%   \centering
%   \includegraphics[width=0.8\linewidth]{../images/wet_lab/tao_g.png}
%   \caption{
%     RNA-seq results of LHX2 overexpressed cells showed that gene ontology (GO) enrichment of differentially expressed genes was mainly concentrated in extracellular matrix related genes.
%   }
%   \label{LHX2_GO}
% \end{figure}

% $CASH$        & Combined Algorithm Selection and Hyperparameter optimization\\
\end{document}
