\documentclass[fleqn,10pt]{wlscirep}
\usepackage[utf8]{inputenc}
\usepackage[T1]{fontenc}
\usepackage{listings}
\usepackage{algorithm2e}
\usepackage{subcaption}
\usepackage[labelformat=parens,labelsep=quad,skip=3pt]{caption}
\usepackage{graphicx}

\title{AGEAS: Automated Machine Learning based Genetic Regulatory Element Extraction System}

\author[1,2,*1,+]{Jack Yu}
% \author[1,2,*1]{Jack Yu}
\author[2,+]{Masayoshi Nakamoto}
\author[1,*2,+]{Jiawang Tao}

% \author[2,+]{Derek Author}
\affil[1]{Center for Health Research, Guangzhou Institutes of Biomedicine and Health, Chinese Academy of Sciences, Guangzhou 510530, China}
\affil[2]{Shenzhen Mozhou Technology Co., Ltd, Shenzhen, China}
% \affil[2]{Affiliation, department, city, postcode, country}
\affil[*1]{Correspondence: gyu17@alumni.jh.edu}
\affil[*2]{Correspondence: tao\_jiawang@gibh.ac.cn}

\affil[+]{these authors contributed equally to this work}
\keywords{Machine Learning, Gene Regulatory Network, RNA-seq, Chip-seq}

\begin{abstract}
  As rapid progress in sequencing technology since last decade, numerous mechanisms underlying cell functions and developmental processes have been revealed as complex regulations of gene expressions.
  Since single-cell RNA sequencing (scRNA-seq) made high-resolution transcriptomic view increasingly accessible, precise identification of gene regulatory network (GRN) describing cell types and cell states became achievable.
  However, extracting key regulatory elements, including gene regulatory pathways (GRPs), transcription factors (TFs), and targetomes, that accurately and completely reflects functionality changes in biological phenomena remains challenging.
  Herein, we describe AGEAS, an semi-supervised automated machine learning (AutoML) based genetic regulatory element extraction system that assesses importances of GRPs in resulting biological phenomena, such as cell type differentiation, physiological and pathological development, and reconstructs GRNs with extracted important GRPs for comprehensive inference.
  With several case studies in divergent research areas, we show that AGEAS can indeed extract informative regulatory elements and reconstruct networks to indicate regulatory changes in biological phenomena of interest.
  % Furthermore, we validated applicability of AGEAS with empirical evidence that LIM homeobox 2 (Lhx2), key quiescent hepatic stellate cell (HSC) factor AGEAS extracted through analyzing chronic carbon tetrachloride ($CCl_4$) induced liver fibrosis data, can indeed halt HSC activation to prevent fibrogenic process.

  \subsection*{Availability and implementation}
  The AGEAS code is available at \url{https://github.com/JackSSK/Ageas}.
\end{abstract}

\begin{document}
\flushbottom
\maketitle

\thispagestyle{empty}
% \noindent Please note: Abbreviations should be introduced at the first mention in the main text â€“ no abbreviations lists. Suggested structure of main text (not enforced) is provided below.

\section*{Introduction}
  \label{introduction}
  As high-resolution sequencing technologies become increasingly applicable and accessible, an efficient and robost analytical system capable of extracting key genetic features responsible for cell types and cell states difference with limited prior biological knowledge also become highly demanded.
  Several methodologies like SCENIC \cite{Scenic} already demonstrated informativeness and robustness in studying cellular phenotypes with GRN analysis.
  Regulons, collections of a TF and corresponding targetomes, can successfully address cell identities with more comprehensive information compared with differential expressing genes (DEGs). \cite{Scenic, pySCENIC}
  Even though limited computational methods are capable of completely and precisly extracting key genetic regulatory elements in biological process of interest, Mogrify\cite{mogrify_2016} and CellNet\cite{cellnet_2014} have demonstrated GRN based methods can help to analyze cell type differentiation and to implement cell reprogramming.
  However, both methods may require large scale of additional background data to be applicable in other studies such as analyzing physiological or pathological development of selected cell type.
  The scREMOTE \cite{scREMOTE} published in 2022 is a potential method to extract generalized regulatory elements since limited background data is required with sequencing data representing sample classes.
  But prior knowledge on key TFs and marker genes is still indispensable.
  Indeed, a regulation trajectory of DEGs or regulons showing significant activity difference after GRN reconstruction would be able to reveal the developmental pathways as generalized approach, but numerous noisy signals are expected due to general gene regulatory nature of TFs.
  Therefore, here we present AGEAS with feasibility in extracting key regulatory elements for biological phenomena of interest even with scarce prior knowledge and potentially promoting further discoveries.

  Several reasons motivated us to develop semi-supervised AutoML based method.
  Firstly, accurate labeling of regulatory elements' relativeness in biological phenomena would require extensive prior knowledge in according research area that may hardly available.
  While unsupervised clustering can potentially extract factos playing important roles, this approach would result in lossing comprehensiveness through focusing on specific aspects with regulatory elements showing similar patterns.
  As a result, we chose to apply biological phenomena reflecting sample classes, such as cell types and cell states which should be easily retrievable, as labels for classification models to predict with masked GRN as input and then interpret success predictions to extract key regulatory elements.
  Secondly, developing a generalized classificaton model in differentiating GRNs is still hard to guarantee comprehensiveness considering that capturing few significant differences would be sufficient for the model to reach outstanding accuracy.
  Hence, a extraction system with multiple independent classification models implemented with various algorithms should be helpful to recover from potential comprehensiveness loss.
  Thirdly, the process of interpreting all success prediction made by every classificaion model and integrating corresponding interpretation results is expected to be extremely computational expensive.
  Thus, applying AutoML-based model and feature selection procedure that can effectively decreas interpretation cost would be necessary to increase feasibility of the method.

  To evaluate performance of AGEAS, we primarily applied AGEAS in studying somatic cell reprogramming process and then further addressed performance in analyzing 3 other biological process in cell subtype differentiation, physiological and pathological development.
  % Later, we also addressed the applicability of AGEAS extraction result in guiding further studies about biological phenomena with $in$ $vitro$ experiments in determining Lhx2, key TF regulating liver fibrosis extracted with AGEAS, can indeed inhibit HSC activation.

\section*{Method}
  \label{method}
  The basic principle of AGEAS is to find key regulatory elements associated with biological phenomena of interest through analyzing how well-performing classification models distinguish GRNs of sample with the phenomena from GRNs of samples without.
  To reconstruct sufficient GRNs for each class, RNA-seq based expression data is segmented into subsets while each one is analogized as a pseudo-sample having discrete expression data.
  The pseudo-sample GRNs (psGRNs) are reconstructed accordingly; thus, classification models can be trained, evaluated, and interpreted having GRPs as input factors.
  With heavily weighted GRPs and correspond genes repeatedly obtained from interpretations of success sample class predictions, GRNs could be formed and potentially playing important role in differentiating the studying sample classes.
  The overall workflow can be summarized as Figure \ref{workflow}.
  By default setting, four seperate extractor units in workflow run parallelly after data preprocessing part in order to increase output stability considering stochastic nature of AGEAS, and all extracted regulatory elements are used to form GRNs later combined into one atlas.
  Following sections describe each step of AGEAS in more depth.
  % \begin{itemize}
  %   \item \hyperref[step1]{\textbf{\emph{Step 1}}}: Data preprocessing
  %   \item \hyperref[step2]{\textbf{\emph{Step 2}}}: Classification model selection.
  %   \item \hyperref[step3]{\textbf{\emph{Step 3}}}: Feature selection...
  %   \item \hyperref[step4]{\textbf{\emph{Step 4}}}: Top GRP extraction.
  %   \item \hyperref[step5]{\textbf{\emph{Step 5}}}: Key network reconstruction.
  % \end{itemize}


  \subsection*{Step 1:Data preprocessing}
    \label{step1}
    The main purpose of this step is to build pseudo-samples and reconstruct corresponding pseudo-sample GRNs (psGRNs).
    For each sample class, gene expression matrices (GEMs) labeled with same class label are concatenated as one comprehensive expression matrix.
    With comprehensive GEMs, a meta-level GRN (meta-GRN) is reconstructed in advance of psGRNs to provide generic guidance on reconstruction.
    In general, the workflow of this step can be represented as Figure \ref{data_preprocess}.

    \subsubsection*{Reconstruct meta-GRN}
      Firstly, genes included in the comprehensive GEMs are assessed and determined whether having potential to form informative GRPs with other genes.
      Commonly, differentially expressed genes (DEGs) would be considered as important factors of studying  phenomena.
      Here we apply the Mann-Whitney U rank test (MWW) implemented by \emph{SciPy}\cite{2020SciPy-NMeth} to exclude genes having indistinguishable expression level distribution across GEMs of different classes.
      The p-value for rejecting null hypothesis, that expression profile underlying class 1 samples is the same as the expression profile underlying class 2 samples, is set to 0.05 by default.
      Furthermore, a $\log_2$ fold change (log2FC) filter is also implemented in AGEAS.
      However, enabling the log2FC filter is not encouraged considering upstream TFs indirectly regulating key genes associated with phenomena of interest may not always have significant differential expression profiles.
      The lof2FC filter shall mostly be used in order to decrease meta-GRN's GRP total amount in a compromising position caused by limited computational resources.
      After differential expression based filters, a standard deviation ($\sigma$) filter is applied to exclude genes either having low expression level or merely affected by dynamic expression status of other genes.
      By default, the $\sigma$ threshold is set to 1.0, the lowest expression level in raw gene count matrix gained from RNA-seq data.
      The threshold value should be adjusted corresponding to prior knowledge of input GEMs, for example of what normalization method was applied to the GEMs.

      With candidate genes passed filters above, some gene pairs are formed and assessed potential of representing GRPs.
      To reduce overall computational complexity, a gene pair shall be formed with at least one TF which could be the regulatory source of GRP.
      If not further specified, TF list will be retrieved from integrated \emph{TRANSFAC}\cite{transfac} datasets according to the provided species information.
      Utilizing genetic interaction database like \emph{GTRD}\cite{gkaa1057} and \emph{BioGRID}\cite{biogrid}, AGEAS checks whether the binding ability of gene pair is confirmed or not.
      By default, if TF being recorded to have binding site within target gene's promoter range (-1000 to +100) by Chromatin Immunoprecipitation Sequencing (ChIP-seq) dataset retrieved from \emph{GTRD}\cite{gkaa1057}, the potential GRP gene pair will be passed to expression correlation assessment.
      For TFs not covered by interaction database, \emph{GRNBoost2}\cite{grnboost2}-like algorithm is initiated to predict potential regulatory target genes.
      The prediction importance threshold can either be set manually or automatically based on recorded interactions as:

      \centerline{$Threshold = min(GA(M_1, t)\, \uplus \, GA(M_2, t), \frac{1}{g})$}

      \noindent Here $GA()$ denotes \emph{GRNBoost2}\cite{grnboost2}-like algorithm;
      $M_1$ is concatenated class 1 samples GEM with genes passed filters;
      $M_2$ is concatenated class 2 samples GEM with genes passed filters;
      $t$ is the TF having largest amount of recorded interaction in dataset;
      $g$ is total amount of unique genes in all samples passed filters.

      After potential GRP gene pairs obtained, a expression correlation filter is used to exclude gene pairs having low covariance which implies weak relationship.
      To assess expression correlation, AGEAS applies Pearson's Correlation coefficient\cite{pcc_2012} (PCC) whihc is one of the widely adopted methods.\cite{cid_2019}
      With default setting, gene pairs can reach absolute correlation coefficient of 0.2 while keeping correspond p-value lower than 0.05 are included in meta-GRN as validated GRP.

    \subsubsection*{Reconstruct psGRNs}
      The comprehensive GEMs is divided into few sample subsets with sliding window algorithm (SWA) to build pseudo-samples.
      With SWA, we can gain GEM of \emph{i-th} pseudo-sample through:

      \centerline{$SWA(i) = \left\{{x_j}_{j = i * p}^{j + k}\right\}, j + k < l$}

      \noindent Here $l$ is number of samples in a comprehensive GEM which can be expressed as $\left\{{x_j}_{j = 0}^{l}\right\}$; $k$ denotes window size; $p$ denotes padding stride.

      If sample amount is considerably low or imbalanced, customized window size and padding stride can be applied to generate sufficient amount of pseudo-samples for later classifier training and assessment processes.
      Utilizing meta-GRN, psGRNs are reconstructed with GEMs of pseudo-sample.
      Each GRP gene pair in meta-GRN is formed with expression data in pseudo-sample and filtered by same PCC filter in meta-GRN reconstruction process.

      After all pseudo-samples have been used to reconstruct psGRN, every psGRN can be represented as matrix comprising GRPs' PCC values.
      The order of GRPs in each psGRN is also unified through adding GRPs included in other psGRNs with 0.0 PCC value.

  \subsection*{Step 2: Classification model selection}
    \label{step2}
    Since AGEAS is not aiming to develop optimized model architectures for psGRN classification but gain insights from multiple models as divergent as possible, the main goal of this step is to select configurations of models capable to make correct psGRN classifications from provided configuration set.
    Regarding the fact that computational power is limited resource, portion of less efficient model configurations shall be pruned although interpreting success predictions of more classification models would lead to more comprehensive insight into sample class differences.
    Therefore, we apply a simple Hyperband\cite{hyperband}-based algorithm \ref{alg:one} which performs grid search for well-performing classification models with varying model training resource and pruning aggressiveness.


    %% This declares a command \Comment
    %% The argument will be surrounded by /* ... */
    \SetKwComment{Comment}{/* }{ */}
    \SetKwInput{kwInput}{Input}
    \SetKwInput{kwReturn}{Return}
    \RestyleAlgo{ruled}
    \begin{algorithm}
    \caption{Model selection algorithm}
    \label{alg:one}
    \kwInput{$R, C, I$ (default $I = 3$), $\alpha_{max}$(default $\alpha_{max} = 0.9$),  $k_{min}$(default $k_{min} = 0.5$)}
    $\alpha_{low} = \frac{1}{(2 ^ {I} - 1)}$\;
    % $X \gets x$\;
    \For{$i \in \{0, 1, ..., I-1\}$}{
      \eIf{$i == I-1$}{
        $\alpha = \alpha_{max}$\;
        % $X \gets X \times X$\;
        % $N \gets \frac{N}{2}$
        % \Comment*[r]{This is a comment}
      }{
        $alpha = 2^i \alpha_{low}$\;
      }
      $r = \alpha R$\;
      $P = \emptyset $\;
      \For{$c \in C$}{
        $p = run\_then\_evaluate(c, r, R)$\;
        Append $p$ to $P$\;
      }
      % \If{$N$ is odd}{
      %   }

      $k = max(1 - \alpha, k_{min}$)\;
  	  $C = top\_configs(C, P, k$)\;
    }
    \kwReturn{$C$}
    \end{algorithm}

    The model selection algorithm requires five inputs
    (1)$R$, the maximum amount of training resource, equivalent to all avaliable psGRNs
    (2)$C$, the total set of provided classification model configurations
    (3)$I$, the number of iterations for model pruning (default set as 3)
    (4)$\alpha_{max}$, the maximum portion of $R$ can be fed to single model (default set as 0.95)
    (5)$k_{min}$, the minimum portion of remaining model configurations will be kept by single pruning iteration (default set as 0.5).
    Furthermore, two functions are also required while need to be defined based on input model configurations:
    \begin{itemize}
    \setlength\itemsep{0em}
    \item \textbf{$run\_then\_evaluate(c, r, R)$}: trains classification model initialized using configuration $c$ for the allocated resource $r$, then returns prediction accuracy (ACC), the area under a receiver operating characteristic curve (AUROC)\cite{hanley_mcneil_1982} score, and total cross-entropy loss ($L_{CE}$) calculated through predicting sample class for all psGRNs $R$.

    \item \textbf{$top\_configs(C, P, k)$}: takes a set of model configurations $C$ with associated evaluation results $P$ and returns configurations having ACC, AUROC score, or $L_{CE}$ reaching top $k$ portion.
    \end{itemize}

    \noindent By default, AGEAS initializes with 128 model configurations utilizing 9 integrated classification algorithms listed in Table \ref{models}.

    \begin{table}[ht]
      \centering
      \begin{tabular}{|l|l|l|l|l|}
      \specialrule{.2em}{.1em}{.1em}
      \textbf{Algorithm} & \textbf{$\# \lambda$} & \textbf{Categorical} & \textbf{Continuous} & \textbf{$\#$ Configs}\\
      \specialrule{.2em}{.1em}{.1em}
      \multicolumn{5}{|l|}{\emph{Implemented with Pytorch}\cite{pytorch}} \\
      \hline
      Transformer & \textbf{14} & 4 & 10 & 32 \\
      \hline
      1D Convolutional Neural Network (1D-CNN) & \textbf{10} & 2 & 8 & 32 \\
      \hline
      Hybrid Convolutional Neural Network (Hybrid-CNN) & \textbf{10} & 2 & 8 & 32 \\
      \hline
      Gated Recurrent Unit (GRU) & \textbf{10} & 4 & 6 & 4 \\
      \hline
      Long Short-Term Memory (LSTM) & \textbf{11} & 4 & 7 & 4 \\
      \hline
      Standard Recurrent Neural Network (RNN) & \textbf{11} & 5 & 6 & 4 \\
      \specialrule{.2em}{.1em}{.1em}
      \multicolumn{5}{|l|}{\emph{Implemented with XGBoost}\cite{chen2016xgboost}} \\
      \hline
      Gradient Boosted Decision Trees (GBDT) & \textbf{18} & 5(1) & 13(4) & 16 \\
      \specialrule{.2em}{.1em}{.1em}
      \multicolumn{5}{|l|}{\emph{Implemented with scikit-learn}\cite{scikit-learn}} \\
      \hline
      Random Forests (RF) & \textbf{14} & 5(1) & 9(1) & 2 \\
      \hline
      Support Vector Machine (SVM) & \textbf{7} & 4 & 3(3) & 2 \\
      \specialrule{.2em}{.1em}{.1em}
      \end{tabular}
      \caption{
        \label{models}
        Classification model algorithms integrated in AGEAS with correspond numbers of hyperparameter and preset model configurations.
        Categorical hyperparameters and continuous numerical hyperparameters are clarifed beside total number of hyperparameters ($\# \lambda$).
        Conditional hyperparameters which are required for selected other hyperparameters are shown in brackets if there is any.
        $\#$ Configs indicates total amount of configurations AGEAS applies by default.
      }
    \end{table}

    The general architecture designs of 1D-CNN and Hybrid-CNN are implemented referring to 1D-CNN and 2D-Hybrid-CNN applied in recent cancer type prediction study.\cite{mostavi_chiu_huang_chen_2020}
    However, taking one convolution layer and adjacent max-pooling layer as a layer set, we implemented both CNN models with flexibilities on number of layer set, which is fixed as 1 in original research.
    An example of 1D-CNN with 2 convolution layer sets is illustrated in Figure \ref{1dCNN}.

    For transformer models, considering psGRNs are already be represented by numerical data matrix while GRPs should barely have positional relationships in the matrix, the embedding layer and positional encoding layer designed for input data tokenization in standard architecture \cite{transformer} are replaced with a single linear layer in AGEAS.

  \subsection*{Step 3: Feature selection}
    \label{step3}
    With selected well-performing models, AGEAS already can start repetition of model training and interpretation in \hyperref[step4]{Step 4} to extract key GRPs.
    However, few uninformative GRPs in training psGRNs merely relied by any model to make classification could be pruned in advance for saving computational power.
    Furthermore, to prevent classification models focusing on a small group of GRPs regardless of training psGRNs, GRPs draw excessive attention shall also be seperated from psGRNs to improve extraction comprehensiveness.
    Thus, in this step, AGEAS iteratively train classificaion models with dynamic $\alpha_{max}$ portion of psGRNs and obtain feature importance scores as described in \hyperref[features_importances]{subsection} below to find GRPs either scored extremely high or considerably low.

    More specifically, at each iteration, GRPs having z-scores ranked as bottom $b$ portion (default set as 0.1) are discarded.
    Also, an $i$-th ranked GRP will be seperated from psGRNs and passed to \hyperref[step5]{Step 5} directly if having z-score fulfilling the condition:

    \centerline{
      $Z_{score}^{i} \ge max(Z_{score}^{thread}, \frac{Z_{score}^{i-1}}{3}, 3 \cdot IQR)$
    }

    \noindent The $Z_{score}^{thread}$ is input z-score threshold (default set as 3.0), and $IQR$ stands for interquartile range calculated by the beginning of each iteration.\
    With this criterion, AGEAS ensures only GRPs draw significantly more attention shall be selected by each iteration despite the data distribution of varying z-score scalled importance values.

    By default, AGEAS iterates this feature selection step for 3 times.

    \subsubsection*{Feature importance estimation}
      \label{features_importances}
      AGEAS applies concept of The Shapley value\cite{roth_1988} for estimating importance of each input feature, equivalent to GRP of input psGRN, in any kind of classification model while making predictions.
      Specific Shapley value calculation or approximating methods are implemented with \emph{SHAP}\cite{lundberg2017unified} and applied to different algorithms as shown in Table \ref{shap}.
      Regarding the standard differences between feature importance estimating methods, we utilize \emph{softmax} function to normalize feature importances and define the normalized importance calculation function as:

      \centerline{$T(X) = softmax(\left\{ F(x) \right\}, x \in X)$}

      \noindent Here $X$ is total set of all features, and $F(x)$ is the importance estimating method of individual classification model being interpreted.\
      If the feature importances can be approached with internalized method $f(x)$, $F(x)$ is set as:

      \centerline{$F(x) = \frac{f(x)}{\sum_{x' \in X} f(x')}$}

      \noindent Otherwise, $F(x)$ is defined utilizing correctly classified input samples $S$, equivalent to psGRNs, with Shapley values $\phi$ of feature $x$ when predicting sample $s$ as class $c_1$ or class $c_2$:\

      \centerline{$F(x) = \sum_{s \in S}\frac{\left|\phi_{c_1,s}^{x}\right| + \left|\phi_{c_2,s}^{x}\right|}{2}$}

      \noindent After all selected classification models $M$ have been interpreted, we can intergrate the feature importances matrices weighted with corresponding models' $L_{CE}$ to one matrix $A$ as:\

      \centerline{
        $ A = \left\{ \sum_{m \in M}(1 - L_{CE}^{m})T_{m}(x) \right\}, x \in X $
      }

      \noindent Then, generalized importance values are obtained through z-score calculation and later sorted by descending order:\

      \centerline{
        $Z_{score} = \left\{\frac{a - \bar{A}}{\sigma_A}\right\}, a \in A$
      }

      \begin{table}[ht]
        \centering
        \begin{tabular}{|l|l|}
        \specialrule{.2em}{.1em}{.1em}
        \textbf{Algorithm} & \textbf{SHAP\cite{lundberg2017unified} Method}\\
        \hline
        Transformer & Gradient Explainer \\
        \hline
        1D-CNN & Deep Explainer \\
        \hline
        Hybrid-CNN & Deep Explainer \\
        \hline
        GRU & Gradient Explainer \\
        \hline
        LSTM & Gradient Explainer \\
        \hline
        RNN & Gradient Explainer \\
        \hline
        GBDT\textbf{*} & Tree Explainer \\
        \hline
        RF & Tree Explainer \\
        \hline
        SVM\textbf{*} & Linear Explainer / Kernel Explainer\\
        \specialrule{.2em}{.1em}{.1em}
        \end{tabular}
        \caption{
          \label{shap}
          Classifier algorithms with applicable Shapley value approximating methods.
          Algorithms marked with \textbf{*} have internalized feature importance estimating methods which will be applied with higher priority than Shapley value based methods.
          GBDT implemented with \emph{XGBoost}\cite{chen2016xgboost} can have feature importance approximated with average weight gain at each split involving the feature.
          Linear SVM implemented with \emph{scikit-learn}\cite{scikit-learn} can have the importance estimated with feature coefficient or using Linear Explainer.
          However, for SVM with kernel function, the feature coefficient estimation would be inappropriate, and feature importances should be approximated with Kernel Explainer.
        }
      \end{table}

  \subsection*{Step 4: Top GRP extraction}
    \label{step4}
    To extract GRPs can effectively define sample class differences, AGEAS iteratively initializes classification models with configurations gained from \hyperref[step2]{Step 2}, trains them with randomly selected $\alpha_{max}$ portion of psGRNs scaled after \hyperref[step3]{Step 3}, and interpret every models' correct predictions as mentioned in \hyperref[features_importances]{subsection} above.
    At each iteration, AGEAS receives a z-score scaled feature importance matrix $A$ and add up each GRP's score accordingly from matrix $A'$ kept from previous iteration if there is one.
    Then, top $a$ (default set as 100) ranked GRPs having z-score greater than $0.0$ extracted from $A$ are compared with GRPs extracted from $A'$ by same setting.
    If less than $d$ (default set as 0.05) portion of GRPs are distinct for GRP sets stratified from $A$ and $A'$, AGEAS will consider the GRP extration result from this iteration being consistent with previous one.
    Extraction iteration will terminate if either encountering $n$ (default set as 3) continuous consistent result or running out of preset iteration number (default set as 10).
    All feature importance scores in matrix $A$ from last iteration are divided by total extraction iteration number processed, and top $a$ ranked GRPs are considered as key GRPs for sample class differentiation and passed to next step.

  \subsection*{Step 5: Key network reconstruction}
    \label{step5}
    Analoging every GRP previously extracted or seperated with high z-score in \hyperref[step3]{Step 3} as a directional edge connecting two gene vertices, equivalent to a TF and a gene, in graph theory, AGEAS attempts to reconstruct a network graph representing regulatory differences between query sample classes.
    Since there is no guarantee on all GRP edges can be connected, some regulatory relationships between gene vertices could be missed.
    Hence, AGEAS utilizes meta-GRN gained from \hyperref[step1]{Step 1} to find GRPs which can further elucidate the regulatory relationships and adds the GRPs back to the network graph.

    For a limited iteration time (default set as  1), AGEAS exhaustively search meta-GRN for TFs which can directly regulate any genes or TFs already covered in the network graph and add the returned TFs as new vertices.
    Next, any GRP in meta-GRN capable to connect two distinct vertices will be added if it is not covered yet.
    The network graph after expansion above represents the key genetic regulatory differences AGEAS extracted from input sample classes.

\section*{Results}
  \label{res}
  To assess predictive power of utilizing extraction result of AGEAS, we first applied AGEAS to study somatic reprogramming to induced pluripotent stem cells (iPSC) achieved in mice by the year 2006, \cite{yamanaka_2006} milestone discovery of cell plasticity. \cite{cell_repro_review}
  With $\sigma$ thread set to 2.0 for constraining total GRP amount in psGRNs, public scRNA-seq based GEMs of embryonic stem cells (ESCs) and mouse embryonic fibroblasts (MEFs) shown in Table \ref{geo_table} were used as input.
  The extraction result can be summarized as TF regulons, and TFs having highest regulatory degree in all extracted GRPs are marked as top TFs (Figure \ref{result_figs}a).

  Among top TFs, Pou5f1 (Oct4) has more association with GRPs extracted as important differential regulatory element between MEF and ESC than any other TFs while forming largest regulon and having expression pattern significantly favoring ESC.
  Thus, we can infer Pou5f1 is key regulatory element associated with ESC identity.
  Extensive studies already addressed Pou5f1's important role in maintaing pluripotency. \cite{niwa_2007, oct4_1, oct4_2}
  Moreover, to implement cell reprogramming to iPSC, Nanog and Sox2 are also noteworthy since they are closely interacting with Pou5f1 in extracted network, having high regulatory degrees, and differentially expressed in ESC.
  Multiple previous studies confirmed that all of Oct4, Nanog, and Sox2 are playing important role to induce cell conversion from somatic cell to iPSC. \cite{yamanaka_2006, ips7f, ipsOK, oct4_nanog_sox2_lin28, oct4_nanog_sox2}
  Cebpb, Hdac2, and Smc3, extracted as top TFs frequently acting as regulatory source in extracted important GRPs and directly affecting Pou5f1 expression, were also reported with relevant functions. \cite{cebpb_ipsc, hdac2_ipsc, rad21_smc3_esc}

  Therefore, we suppose investigating TFs acting as common regulatory sources in extracted important differential GRPs and correspond regulons can effectively study differences between cell types cell states.
  % \subsection*{In silico evaluations}
  To further address AGEAS's applicability and limitation, we applied AGEAS to three other scRNA-seq based studies in distinct research areas.
  With all default settings describe in \hyperref[method]{Method} section above, we can assess AGEAS's performance in following scenarios with public data set (Table \ref{geo_table}):

  \begin{itemize}
      \setlength\itemsep{0em}
      \item {\textbf{Dopaminergic neuron generation}}:

        Analyzing difference between human iPSC-derived radial glial / neuronal co-culture, as neural progenitors,\cite{ASCL1_dopaminergic_neuron_2021} and tyrosine hydroxylase (TH) expressing purified dopaminergic neurons (DANs), we address the applicability of AGEAS on cell subtype differentiation problems.

        From extracted TFs (Figure \ref{result_figs}b), we hypothesize ISL1, ELF3, and PBX3 playing important role in DAN differentiation due to their high regulatory degree in extracted important GRPs.
        Previously, ISL1 was determined essential for differentiation of prethalamic DANs, \cite{isl1_da} and ELF3 was found to be neuronal precursor cell marker associated with neural stem cell development, \cite{ELF3_precursor_marker} consistent with it's high expression level in neuronal co-culture samples.
        Moreover, a DAN development study reported PBX3 being required for correct differentiation of neuroblast, product of radial glial cell, into midbrain DAN and survival of midbrain DAN. \cite{pbx3_dan}

        Through substratification of all extracted GRPs, we only found HDAC2 and RBPJ as potential upstream regulatory elements of all ISL1, ELF3, and PBX3.
        Few reports demonstrated HDAC2 is associated with neurogenesis of radial glial cells. \cite{hdac2_dan_1, hdac2_dan_2}
        Nevertheless, HDAC2's role in DAN differentiation is yet to be clarified.
        RBPJ was reported essential for DAN survival and affecting DAN development by regulating ASCL1 which is an essential factor in neurogenesis. \cite{rbpj_1, rbpj_2}
        Original research where we retrieved scRNA-seq data also demonstrated ASCL1 as necessary factor regulating DA neurotransmitter selection. \cite{ASCL1_dopaminergic_neuron_2021}
        Although AGEAS extracts ASCL1 as one of top TFs; however, no direct regulatory relationship between RBPJ and ASCL1 is identified.
        % PHOX2B, and ELF3 are regulating essential functions differentiating DANs from neural progenitors referring to their correspond regulon sizes and differential expression profiles.
        % Both ISL1 and PHOX2B were found related with DA phenotype in previous studies:
        % ISL1 is
        % and PHOX2B is connected with TH expression influencing DAN development and improving DA activity. \cite{phox2_caudal_da, phox2_rat_da}

        % EBF1, which forms the largest regulon with extracted GRPs, was determined as key neuronal differentiation regulator in the ventral telencephalon, nature source of neuronal and glial cells. \cite{ebf1_striatum, ebf1_cell_diff}
        % Furthermore, EBF1 plays important role in mesodiencephalic DAN migration. \cite{ebf1_migration}


      \item {\textbf{Postnatal cardiomyocyte maturation}}:

        To address the performance of AGEAS while analyzing cell physiological development, scRNA-seq data of cardiomyocytes (CMs) in postnatal day 7 mice (P7) and day 28 mice (P28) are used as input for extracting key genetic regulatory elements in postnatal cardiomyocyte maturation.

        As indicated by extracted differential GRPs, we primarily investigated Sp1 and Esr1 ($ER\alpha$) which also have top regulatory degrees and considerably expression changes (Figure \ref{result_figs}c).
        Previous studies demonstrated that Sp1 promotes CM hypertrophy \cite{sp1_hypertrophy} and maturation of electrophysiology, $Ca^{2+}$ handling. \cite{Sp1_electrophysiologt, CM_mature}
        % trans-activation of the SERCA2 promoter
        Esr1 was found modulating myocardial development for postnatal cardiac growth.\cite{esr1_cm, esr1_cm_growth}

        Beside inter-regulation between Sp1 and Esr1, 3 other top TFs (Srf, Cebpb, Rest) and 2 TFs forming relatively small regulons (Gata4, Prox1), are expected to have reversible regulation with both Sp1 and Esr1 based on extraction network.
        All 5 TFs were found having CM maturation related functions. (Table \ref{interact_TFs})

        \begin{table}[ht]
            \centering
            \begin{tabular}{|l|l|}
              \hline
              \textbf{Gene} & \textbf{Function}  \\
              \hline
              Srf & Regulate sarcomere genes and broadly impact almost every aspect of CM maturation\cite{CM_mature} \\
              \hline
              Cebpb & Repress CM proliferation and hypertrophy\cite{cebpb_1, cebpb_2}\\
              \hline
              Rest (Nrsf) & Repress CM hypertrophy; Prevent dilated cardiomyopathy\cite{rest_1}\\
              \hline
              Prox1 & Repress CM hypertrophy; Prevent dilated cardiomyopathy\cite{prox1_CM} \\
              \hline
              Gata4 & Promote CM proliferation and hypertrophy\cite{CM_posnatal_mature} \\
              \hline
            \end{tabular}
            \caption{
              \label{interact_TFs}
              TFs interacting with both Sp1 and Esr1 order by extracted regulon size.
            }
        \end{table}

        % Other top TFs expected to directly influence expression of either Sp1 or Esr1 are:
        % \begin{table}[ht]
        %     \centering
        %     \begin{tabular}{|l|l|}
        %       \hline
        %       \textbf{Gene} & \textbf{Function}  \\
        %       \hline
        %       Nfe2l2 &  mitochondrial and cardiac oxidative stress, cardiac fibrosis (10.1089/ars.2016.6831) \\
        %       \hline
        %       Nfib & 1 \\
        %       \hline
        %       Cbfb & 1 \\
        %       \hline
        %       Fosl2 & 1 \\
        %       \hline
        %       Nkx2-5 & \cite{cebpb_1, nkx25, CM_posnatal_mature} \\
        %       \hline
        %       Rxra & 1 \\
        %       \hline
        %       Mef2a & 1 \\
        %       \hline
        %       Yap1 & \cite{CM_mature, yap1, CM_posnatal_mature} \\
        %       \hline
        %       Thra & 1 \\
        %       \hline
        %     \end{tabular}
        %     \caption{
        %       \label{interact_TFs}
        %       Extracted top TFs interacting with either Sp1 or Esr1.
        %     }
        % \end{table}


      \item {\textbf{Hepatic stellate cell activation in $CCl_4$ induced liver fibrosis}}:

        Here we address AGEAS's applicability on cell pathological development studies through analyzing hepatic stellate cells (HSCs) in mice liver administrated with chronic carbon tetrachloride ($CCl_4$) for 6 weeks and according portal fibroblasts (PF) simulating activated HSCs.

        Based on extracted important GRPs, we identified Mef2c, Jun, and Jund as potential representitives of key regulatory elements in HSC activation (Figure \ref{result_figs}d).
        As TF extracted with most important GRPs, Mef2c has been well-documented as key regulator in HSC activation. \cite{mef2c_1, mef2c_2, mef2c_3}
        Extensive studies also found Jun and Jund, both being functional components of the AP1 TF complex, essential for fibrosis-related HSC activation and profibrogenic process. \cite{ap1_hsc_1, ap1_hsc_2, ap1_hsc_3, ap1_hsc_4}

        Analyzing TF regulons of Mef2c, Jun, and Jund, we suppose Mef2c is acting as targetome of Jund and inter-regulatory element of Jun.
        While no regulatory relationship between Jund and Jun determined, 3 extracted top TFs (Lhx2, Nfib, Junb) can potentially influence expressions of both Jund and Jun.
        Since Junb is also functional component of AP1, we infer it associated with HSC activaion in a similar manner.
        Furthermore, Lhx2 was found indispensable for quiescent HSCs. \cite{lhx2_fibro, Lhx2_hsc_1}
        Although, little literature addressed Nfib's role in HSC, Nfib was reported acting as anti-apoptotic gene for cell proliferation during $CCl_4$ induced liver damage. \cite{Nfib_sadly}
    \end{itemize}

  % \subsection*{In vitro validation}
  %   To demonstrate empirically the predictive capabilities of utilizing AGEAS extraction result, we further investigated the extraction result of $CCl_4$ induced liver fibrosis study setting scenario of identifying theraputic targets inhibiting pathological fibrogenic process in liver.
  %   Through analyzing extracted TF regulons, we further confirmed Mef2c's role as key targetome noticing direct regulational relationships with HSC activation marker Acta2\cite{acta2_hsc_act} and fibrotic collagen Col1a1 (Figure \ref{regulons_figs}d).
  %   Moreover, since only Lhx2 has all related GRPs mainly being functional in HSC (Figure \ref{regulons_figs}a) and significantly differential expression in HSC (Figure \ref{result_figs}d), we decided to proceed manipulation of Lhx2 expression.
  %   Previous study already proved knockdown of Lhx2 promotes liver fibrosis in mice.\cite{lhx2_fibro}
  %   Thus, we suppose overexpression of Lhx2 could assist to maintain quiescent of HSCs through Jun/Jund-Mef2c dependent pathways.
  %
  %   To verify our hypothesis, we firstly validated the association between Lhx2 and HSC activation with $in$ $vitro$ experiments.
  %   In mice primary HSC (mHSC) and human HSC cell lines (LX2), we detected the expression of Lhx2/LHX2 with the activation of HSC.
  %   The results showed that the expression of Lhx2 was significantly decreased with the activation of HSC (Figure \ref{hsc_activation_mRNA}), consistent with our previous analysis result.
  %   To further clarify the role of LHX2 in human cell, we knocked down LHX2 by small interfering RNA (siRNA) in LX2 cells.
  %   The results showed that knockdown of LHX2 significantly promoted the expression of fibrosis-related genes in Figure \ref{hsc_activation_mRNA}d, compatible with previous knowledge in mice.
  %
  %   Subsequently, we overexpressed Lhx2/LHX2 in mHSCs and LX2 cells.
  %   The results showed that overexpression of LHX2 directly affected the cell morphology of human HSCs and significantly inhibited the expression of fibrogenic genes (Figure \ref{LHX2_LX2_morphology} and \ref{LHX2_OEs}a).
  %   In addition, observation in mHSCs further confirmed that Lhx2 could inhibit the expression of fibrosis-related genes (Figure \ref{LHX2_OEs}b).
  %   In order to comprehensively understand the gene expression profile of HSC regulated by LHX2, RNA-seq results showed that the differential genes were mainly related to extracellular matrix structure and assembly (Figure \ref{LHX2_GO}), which was consistent with fibrosis-related phenotype.
  %
  %   Thus, systematic and comprehensive $in$ $vitro$ verification showed that Lhx2/LHX2 is indeed capable to regulate fibrogenic gene expressions and inhibit HSC activation as we concluded from AGEAS extraction output.
  %   However, the preliminarily identified regulatory pathways for Lhx2 to inhibit HSC activation is yet to be addressed.
  %   % In vivo experiment will take longer time...

\section*{Conclusion}
  With result achieved so far, AGEAS shows robustness in extracting key regulatory elements for several biological researches.
  Key TFs inferred by extracted important GRPs and regulon analysis in all 4 $in$ $silico$ cases were confirmed informative to reveal main functionalities responsible for biological process of interest.
  Furthermore, the direct extractions of ISL1 in DAN differentiation and Jund in HSC activation as regulatory source in top amount of important GRPs can demonstrate that AGEAS is not focusing solely on TF regulon sizes in reconstructed meta-GRN nor expression differences between sample classes.
  % Also, $in$ $vitro$ experiments indicate great potential in AGEAS to promote method development in phenotypic manipulation.
  % While TF regulons reconstructed with all extracted GRPs are having sizes proportional to meta-GRN... GO enrichment?
  % However, whether extracted regulatory elements covering all functional aspects of studying biological phenomena, admittedly, remains undetermined due to lack of validation for GRPs' activities.
  Overall, we anticipate that AGEAS will become useful in providing insightful regulatory information and promoting pioneer researches in revealing complicate mechanism behind biological phenomena.


\section*{Discussion}
  \label{disc}
  Considering the rapid development of both sequencing technology and machine learning algorithms, we implemented
  AGEAS with modular design.
  In data preprocessing part, the meta-GRN reconstruction section can be replaced with other methods to further increase fidelity of regulatory relationships or integrating more information such as cis-regulatory elements (CREs) accessibility which can be obtained from Assay for Transposase-Accessible Chromatin Sequencing (ATAC-seq) data.
  Moreover, CRE accessibility information combining with motif enrichment analysis is proved capable to identity potential regulatory pathways replacing ChIP-seq based dataset AGEAS utilizing in present study. \cite{scREMOTE}
  Refine meta-GRN reconstruction method in consideration of having input data from simultaneous scRNA-seq and single cell ATAC-seq (scATAC-seq), such as SHARE-seq published in late 2020, \cite{share_seq} can potentially further improve applicability of AGEAS; however, accessibility of corresponding datasets will be essential for performance assessment and method development.
  Similarly, initial set of classificaion models is also subject to be changed responding to future computational studies.

  Also, it is important to note that AGEAS can be applied with GEMs normalized with different methods on user's choice as well as derived from other sequencing technologies such as bulk RNA-seq or spatial sequencing in revealing tissue differences.
  Replacing transcriptiomic data with proteomic data may also be achievable to analysis protein interaction networks.
  However, larger scale of applicability tests shall be performed in advance.

  Due to the stochastic nature of AGEAS, extraction results of AGEAS are not expected to stay identical for repeated applications with exactly same inputs.
  Main reason underlying the inconsistency is caused by how AGEAS attempt to obtain top performing models for later prediction interpretations.
  Repeated random subsets of psGRNs using in model training and selection are expected to vary for each application hence resulting in different performance of classification models and ranking results.
  As a result, the importance values of GRPs might differ and slightly affect top TFs extracted later.

  Two possible workarounds of inconsistent issue would be:
  (i) manually set random seed for every process influencing subset selection.
  (ii) set multiple extractor unit and combine extraction results.
  In present study, we found n = 4 units yielding relatively stable results, but this value could vary based on application scenario.

  % Although Lhx2 extracted by AGEAS-based analysis indeed shows capability to inhibit HSC activation during liver fibrosis as predicted, the corresponding inference of functional pathways still need to be proved.
  % Furthermore, $in$ $vivo$ experiments are also necessary to evaluate functionality of Lhx2 and associated pathways for higher fidelity.

\bibliography{reference}

% For data citations of datasets uploaded to e.g. \emph{figshare}, please use the \verb|howpublished| option in the bib entry to specify the platform and the link, as in the \verb|Hao:gidmaps:2014| example in the sample bibliography file.

\section*{Author contributions statement}
  \textbf{J.Y.}: Methodology, Software, Writing- Original draft preparation, Project administration
  \textbf{M.N.}: Methodology, Software
  \textbf{J.T.}: Methodology, Writing- Original draft preparation, Project administration
% Writing - Review \& Editing,

\section*{Additional information}
  All scRNA-seq datasets are retrieved from Gene Expression Omnibus(GEO) as described in the Table below.

  \begin{table}[ht]
    \centering
    \begin{tabular}{|l|l|}
    \specialrule{.2em}{.1em}{.1em}
    \textbf{Sample Class} & \textbf{Accession number}\\

    \specialrule{.2em}{.1em}{.1em}
    \multicolumn{2}{|l|}{GSE103221} \\
    \hline
    MEF & GSM3629847 \\
    \hline
    ESC & GSM3629848 \\

    \specialrule{.2em}{.1em}{.1em}
    \multicolumn{2}{|l|}{GSE137720 } \\
    \hline
    $CCl_4$ a6w hsc & GSM4085625 \\
    \hline
    $CCl_4$ a6w pf & GSM4085627 \\
    % \hline
    % healthy hsc & GSM4085623 \\
    % \hline
    % healthy pf & GSM4085626 \\

    \specialrule{.2em}{.1em}{.1em}
    \multicolumn{2}{|l|}{GSE185275 } \\
    \hline
    glial/neuronal co-culture & GSM5609927 \\
    \hline
    purified DANs & GSM5609930 \\

    \specialrule{.2em}{.1em}{.1em}
    \multicolumn{2}{|l|}{GSE156482 } \\
    % \hline
    % p4 cm & GSM4732219 \\
    \hline
    p7 CM & GSM4732221 \\
    \hline
    p28 CM & GSM4732225 \\
    % \hline
    % p28 ncm & GSM4732226 \\

    \specialrule{.2em}{.1em}{.1em}
    \end{tabular}
    \caption{
      \label{geo_table}
    }
  \end{table}

\begin{figure}[ht]
  \centering
  % \includegraphics[width=0.8\linewidth, height=18cm, keepaspectratio,]{../images/summary_trans.png}
  \includegraphics[width=0.8\linewidth, keepaspectratio,]{../images/summary_trans.png}
  \caption{
    The overall workflow of AGEAS:
    \textbf{\emph{(1)}} Reconstruct meta-level GRN (meta-GRN) with expression data of all samples.
    \textbf{\emph{(2)}} Build pseudo-samples with sliding window algorithm and reconstruct GRNs with GRPs identified in meta-GRN accordingly.
    \textbf{\emph{(3)}} Select best performing classifiers in predicting class labels of pseudo-sample GRNs (psGRNs).
    \textbf{\emph{(4)}} Interpret how top models make classifications and gradually exclude GRPs with low weights or outlier-level high weights.
    \textbf{\emph{(5)}} Repeatedly train classifiers with different set of psGRNs as training data to extract GRPs frequently ranked as top important features for decision.
    \textbf{\emph{(6)}} Reconstruct GRNs with extracted GRPs and GRPs excluded as significant outliers.
  }
  \label{workflow}
\end{figure}

\begin{figure}[ht]
  \centering
  % \includegraphics[width=0.8\linewidth, height=12cm, keepaspectratio,]{../images/data_preprocess_trans.png}
  \includegraphics[width=0.8\linewidth, keepaspectratio,]{../images/data_preprocess_trans.png}
  \caption{
    Workflow to reconstruct meta-GRN and psGRNs.
    \textbf{\emph{(1)}} Filter genes from GEMs with log2FC filter(optional), MWW filter, and $\sigma$ filter.
    \textbf{\emph{(2)}} Find candidate GRP gene pairs from either interaction database or predictions made by \emph{GRNBoost2}\cite{grnboost2}-like algorithm.
    \textbf{\emph{(3)}} Filter candidate GRP gene pairs with PCC filter and reconstruct meta-GRN with validated GRPs.
    \textbf{\emph{(4)}} Generate pseudo-samples with SWA.
    \textbf{\emph{(5)}} Utilize GRPs in meta-GRN as generic guidance to form candidate GRPs for pseudo-samples.
    \textbf{\emph{(6)}} Filter every candidate GRP for all pseudo-samples with PCC filter and reconstruct psGRNs with validated GRPs.
  }
  \label{data_preprocess}
\end{figure}

\begin{figure}[ht]
  \centering
  \includegraphics[width=0.8\linewidth]{../images/nn.png}
  \caption{1D-CNN with 2 convolution layer set.}
  \label{1dCNN}
\end{figure}

\begin{figure}
  \begin{subfigure}{0.49\linewidth}
    \centering
    \includegraphics[width=\linewidth, keepaspectratio,]{../images/top_TFs/MEF_ESC.png}
    \caption{}
  \end{subfigure}
  \begin{subfigure}{0.49\linewidth}
    \centering
    \includegraphics[width=\linewidth, keepaspectratio,]{../images/top_TFs/N_cocul.png}
    \caption{}
  \end{subfigure}

  \begin{subfigure}{0.49\linewidth}
    \centering
    \includegraphics[width=\linewidth, keepaspectratio,]{../images/top_TFs/CMp7d_CMp28d.png}
    \caption{}
  \end{subfigure}
  \begin{subfigure}{0.49\linewidth}
    \centering
    \includegraphics[width=\linewidth, keepaspectratio,]{../images/top_TFs/HSCp6w_PFp6w.png}
    \caption{}
  \end{subfigure}

  \caption{
    Top 15 TFs ranked by regulatory degrees in all GRPs extracted by AGEAS.
    % Heatmap at right indicates relative $log_2$ gene expression levels in binary sample classes.
    \textbf{\emph{(a)}} Mouse embryonic fibroblast vs. Embryonic stem cell
    \textbf{\emph{(b)}} Purified dopaminergic neuron vs. Radial glial/neuronal co-culture
    \textbf{\emph{(c)}} 7 days postnatal cardiomyocyte vs. 28 days postnatal cardiomyocyte
    \textbf{\emph{(d)}} Hepatic stellate cell vs. Portal fibroblast (both after 6 weeks of CCl4 administration)
  }
  \label{result_figs}
\end{figure}

% \begin{figure}
%
%   \begin{subfigure}{0.49\linewidth}
%     \centering
%     \includegraphics[width=\linewidth, keepaspectratio,]{../images/lhx2_class.png}
%     \caption{Lhx2 regulon}
%   \end{subfigure}
%   \begin{subfigure}{0.49\linewidth}
%     \centering
%     \includegraphics[width=\linewidth, keepaspectratio,]{../images/junb_class.png}
%     \caption{Junb regulon}
%   \end{subfigure}
%
%   \begin{subfigure}{0.49\linewidth}
%     \centering
%     \includegraphics[width=\linewidth, keepaspectratio,]{../images/nfib_class.png}
%     \caption{Nfib regulon}
%   \end{subfigure}
%   \begin{subfigure}{0.49\linewidth}
%     \centering
%     \includegraphics[width=\linewidth, keepaspectratio,]{../images/mef2c_class.png}
%     \caption{Mef2c regulon}
%   \end{subfigure}
%
%   \caption{
%     Regulons of extracted top TFs interacting with both Jund and Jun and Mef2c.
%     Class 1 stands for HSC; class 2 stands for PF.
%   }
%   \label{regulons_figs}
% \end{figure}

% \begin{figure}
%   \begin{subfigure}{0.49\linewidth}
%     \centering
%     \includegraphics[width=\linewidth, keepaspectratio,]{../images/wet_lab/tao_a.png}
%     \caption{}
%   \end{subfigure}
%   \begin{subfigure}{0.49\linewidth}
%     \centering
%     \includegraphics[width=\linewidth, keepaspectratio,]{../images/wet_lab/tao_b.png}
%     \caption{}
%   \end{subfigure}
%
%   \begin{subfigure}{0.49\linewidth}
%     \centering
%     \includegraphics[width=\linewidth, keepaspectratio,]{../images/wet_lab/tao_c.png}
%     \caption{}
%   \end{subfigure}
%   \begin{subfigure}{0.49\linewidth}
%     \centering
%     \includegraphics[width=\linewidth, keepaspectratio,]{../images/wet_lab/tao_h.png}
%     \caption{}
%   \end{subfigure}
%   \caption{
%     Relative mRNA expression levels during HSC activation.
%     \textbf{\emph{(a)}} Fibrosis-related gene (Col1a1, Col3a1, Acta2) expression in mHSCs were increased with the extension of culture time, indicating HSC activation.
%     n = 3.
%     \textbf{\emph{(b)}} With the activation of mHSC, expression of Lhx2 decreased gradually.
%     n = 3.
%     \textbf{\emph{(c)}} In human HSC cell line LX2, TGF-$\beta$ can remarkably activate HSC and promote expressions of fibrosis-related genes. % Reference ?
%     Meanwhile, the expression of LHX2 decreased with the activation of HSC.
%     \textbf{\emph{(d)}} Knockdown of LHX2 by siRNA (si-LHX2) significantly increased the expression of fibrosis-related genes in LX2 cells compared with negative control (si-NC).
%     n = 4.
%   }
%   \label{hsc_activation_mRNA}
% \end{figure}
%
% \begin{figure}
%   \begin{subfigure}{0.49\linewidth}
%     \centering
%     \includegraphics[width=\linewidth, keepaspectratio,]{../images/wet_lab/tao_d_lx2_wt.png}
%     \caption{LX2-WT}
%   \end{subfigure}
%   \begin{subfigure}{0.49\linewidth}
%     \centering
%     \includegraphics[width=\linewidth, keepaspectratio,]{../images/wet_lab/tao_d_lx2_pcdh_oe.png}
%     \caption{LX2-PCDH-OE}
%   \end{subfigure}
%
%   \begin{subfigure}{0.49\linewidth}
%     \centering
%     \includegraphics[width=\linewidth, keepaspectratio,]{../images/wet_lab/tao_d_lx2_lhx2_oe.png}
%     \caption{LX2-LHX2-OE}
%   \end{subfigure}
%   \caption{
%     Overexpression of LHX2 significantly altered cell morphology in LX2 cell lines.
%     \textbf{\emph{(a)}} LX2 wild type.
%     \textbf{\emph{(b)}} LX2 overexpressing negative control vector.
%     \textbf{\emph{(c)}} LX2 overexpressing LHX2.
%   }
%   \label{LHX2_LX2_morphology}
% \end{figure}
%
% \begin{figure}
%   \begin{subfigure}{0.49\linewidth}
%     \centering
%     \includegraphics[width=\linewidth, keepaspectratio,]{../images/wet_lab/tao_e.png}
%     \caption{LX2}
%   \end{subfigure}
%   \begin{subfigure}{0.49\linewidth}
%     \centering
%     \includegraphics[width=\linewidth, keepaspectratio,]{../images/wet_lab/tao_f.png}
%     \caption{mHSC}
%   \end{subfigure}
%
%   \caption{
%     Overexpression of Lhx2/LHX2 significantly decreased the expression of fibrosis-related genes.
%     \textbf{\emph{(a)}} in LX2 cells. n = 3.
%     \textbf{\emph{(b)}} in mHSCs. n = 3.
%   }
%   \label{LHX2_OEs}
% \end{figure}
%
% \begin{figure}[ht]
%   \centering
%   \includegraphics[width=0.8\linewidth]{../images/wet_lab/tao_g.png}
%   \caption{
%     RNA-seq results of LHX2 overexpressed cells showed that gene ontology (GO) enrichment of differentially expressed genes was mainly concentrated in extracellular matrix related genes.
%   }
%   \label{LHX2_GO}
% \end{figure}

% $CASH$        & Combined Algorithm Selection and Hyperparameter optimization\\
\end{document}
